{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d9f808",
   "metadata": {},
   "source": [
    "# Preprocess Gene Annotations\n",
    "\n",
    "This notebook creates a table of gene annotations by:\n",
    "1. Querying Biomart for all Ensembl IDs in the database\n",
    "2. Querying MyGene for annotation about those IDs\n",
    "3. Querying Ensembl for the most recent Ensembl release for each ID\n",
    "4. Building a permalink to the Ensembl archive page for each ID\n",
    "\n",
    "This gene annotation table is read in by `agoradataprocessing/process.py` to be used in the `gene_info` transformation. \n",
    "\n",
    "***Note:*** *This notebook is exploratory and should eventually be converted to a Python script that is run through an automated process.*\n",
    "\n",
    "## Installation requirements\n",
    "\n",
    "#### Linux / Windows / Mac\n",
    "\n",
    "Install R: https://cran.r-project.org/\n",
    "\n",
    "Install Python and agora-data-tools following the instructions in this repository's README. This notebook assumes it is being run from the same `pipenv` virtual environment as agora-data-tools. \n",
    "\n",
    "Then install the following packages using `pip`:\n",
    "```\n",
    "pip install rpy2 mygene\n",
    "```\n",
    "\n",
    "#### Note for Macs with M1 chips (2020 and newer)\n",
    "\n",
    "Install as above, but make sure that your R installation is the arm64 version (R-4.X.X-arm64.pkg) so that the architecture matches what pip is using. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2dc92c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocManager\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.3 (2023-03-15)\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Bioconductor version '3.15' is out-of-date; the current release version '3.16'\n",
      "  is available with R version '4.2'; see https://bioconductor.org/install\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: biomaRt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from rpy2.robjects import r\n",
    "from os import name\n",
    "import pandas as pd\n",
    "import mygene\n",
    "import numpy as np\n",
    "import requests\n",
    "import synapseclient\n",
    "import agoradatatools.etl.transform as transform\n",
    "import agoradatatools.etl.utils as utils\n",
    "import agoradatatools.etl.extract as extract\n",
    "\n",
    "r('if (!require(\"BiocManager\", character.only = TRUE)) { install.packages(\"BiocManager\") }')\n",
    "r('if (!require(\"biomaRt\")) { BiocManager::install(\"biomaRt\") }')\n",
    "\n",
    "r.library('biomaRt')\n",
    "\n",
    "biomart_filename = '../output/biomart_ensg_list.txt'\n",
    "archive_filename = '../output/ensembl_archive_list.csv'\n",
    "config_filename = '../../config.yaml'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec36f54c",
   "metadata": {},
   "source": [
    "# Part 1: Get gene annotation data\n",
    "\n",
    "## Query Biomart for a list of all Ensembl IDs in the database of human genes. \n",
    "\n",
    "Here we use the R library `biomaRt`. There is no canonical Python library with the features we need for this script. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d1e115d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460']\n",
      "69299 genes found.\n"
     ]
    }
   ],
   "source": [
    "# Sometimes Biomart doesn't respond and the command needs to be sent again. Try up to 5 times.\n",
    "for T in range(5):\n",
    "    try:\n",
    "        mart = r.useEnsembl(biomart = 'genes', dataset = 'hsapiens_gene_ensembl')\n",
    "        ensembl_ids = r.getBM(attributes = 'ensembl_gene_id', mart = mart, useCache = False)\n",
    "        \n",
    "    except rpy2.rinterface.RRuntimeError as err:\n",
    "        print(err)\n",
    "        print('Trying again...')\n",
    "    \n",
    "    else: \n",
    "        break\n",
    "\n",
    "if ensembl_ids.nrow == 0:\n",
    "    print('Biomart was unresponsive after 5 attempts. Try again later.')\n",
    "\n",
    "else:\n",
    "    # Convert the ensembl_gene_id column from R object to a python list\n",
    "    ensembl_ids = list(ensembl_ids.rx2('ensembl_gene_id'))\n",
    "\n",
    "    print(ensembl_ids[0:5])\n",
    "    print(str(len(ensembl_ids)) + \" genes found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f494e58f",
   "metadata": {},
   "source": [
    "## Add Ensembl IDs from data sets that will be processed by agora-data-tools\n",
    "\n",
    "Some of these datasets have older/retired Ensembl IDs that no longer exist in the current Ensembl database. Loop through all data sets in the config file to get all Ensembl IDs they use, and add missing ones to the gene list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3fdbeec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes_biodomains': ('syn44151254.1', 'csv'),\n",
       " 'neuropath_regression_results': ('syn22017882.5', 'csv'),\n",
       " 'agora_proteomics': ('syn18689335.3', 'csv'),\n",
       " 'agora_proteomics_tmt': ('syn35221005.2', 'csv'),\n",
       " 'target_exp_validation_harmonized': ('syn24184512.6', 'csv'),\n",
       " 'metabolomics': ('syn26064497.1', 'feather'),\n",
       " 'igap': ('syn12514826.4', 'csv'),\n",
       " 'eqtl': ('syn12514912.3', 'csv'),\n",
       " 'proteomics': ('syn18689335.3', 'csv'),\n",
       " 'rna_expression_change': ('syn27211942.1', 'tsv'),\n",
       " 'target_list': ('syn12540368.36', 'csv'),\n",
       " 'median_expression': ('syn27211878.2', 'csv'),\n",
       " 'druggability': ('syn13363443.11', 'csv'),\n",
       " 'team_info': ('syn12615624.13', 'csv'),\n",
       " 'team_member_info': ('syn12615633.15', 'csv'),\n",
       " 'overall_scores': ('syn25575156.13', 'table'),\n",
       " 'networks': ('syn11685347.1', 'csv'),\n",
       " 'diff_exp_data': ('syn27211942.1', 'tsv'),\n",
       " 'proteomics_tmt': ('syn35221005.2', 'csv')}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = utils._get_config(config_path = config_filename)\n",
    "datasets = utils._find_config_by_name(config, 'datasets')\n",
    "\n",
    "files = {}\n",
    "\n",
    "for dataset in datasets:\n",
    "    dataset_name = list(dataset.keys())[0]\n",
    "    \n",
    "    for entity in dataset[dataset_name]['files']:\n",
    "        entity_id = entity['id']\n",
    "        entity_format = entity['format']\n",
    "        entity_name = entity['name']\n",
    "        \n",
    "        # Ignore json files, which are post-processed and not what we're interested in. \n",
    "        # Also ignore \"gene_metadata\" since that's the file we're making here.\n",
    "        if entity_format != 'json' and entity_name != \"gene_metadata\":\n",
    "            files[entity_name] = (entity_id, entity_format)\n",
    "\n",
    "# There are some duplicate synID's in this list but that doesn't really matter\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1a2120",
   "metadata": {},
   "source": [
    "### We should now have a list of all raw data files ingested. Get each one and add its Ensembl IDs to the gene list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9843689d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:synapseclient_default:Welcome, Jaclyn Beck!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target_exp_validation_harmonized has an n/A Ensembl ID\n"
     ]
    }
   ],
   "source": [
    "syn = utils._login_to_synapse(token = None) # Assumes you have already logged in with a valid token\n",
    "\n",
    "# The various column names used to store Ensembl IDs in the files\n",
    "col_names = ['ENSG', 'ensembl_gene_id', 'GeneID']\n",
    "\n",
    "for file in files.keys():\n",
    "    df = extract.get_entity_as_df(syn_id=files[file][0],\n",
    "                                  source=files[file][1],\n",
    "                                  syn=syn)\n",
    "    file_ensembl_ids = None\n",
    "    \n",
    "    for C in col_names:\n",
    "        if C in df.columns:\n",
    "            file_ensembl_ids = df[C]\n",
    "    \n",
    "    # networks file is a special case\n",
    "    if file == \"networks\":\n",
    "        file_ensembl_ids = pd.melt(df[['geneA_ensembl_gene_id','geneB_ensembl_gene_id']])['value']\n",
    "        \n",
    "    if file_ensembl_ids is not None:\n",
    "        ensembl_ids = ensembl_ids + file_ensembl_ids.tolist()\n",
    "        if \"n/A\" in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an n/A Ensembl ID\")\n",
    "        if np.NaN in file_ensembl_ids.tolist():\n",
    "            print(file + \" has an NaN Ensembl ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d66ddf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70446"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_ids = list(set(ensembl_ids))\n",
    "ensembl_ids.remove(\"n/A\") # Necessary because one data set has an Ensembl ID set to \"n/A\"\n",
    "\n",
    "# Convert to a pandas data frame\n",
    "ensembl_ids_df = pd.DataFrame({'ensembl_gene_id': ensembl_ids})\n",
    "len(ensembl_ids_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4e7a37c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to a file\n",
    "ensembl_ids_df.to_csv(path_or_buf = biomart_filename, sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13ef6ec",
   "metadata": {},
   "source": [
    "## Get info on each gene from mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7ebd03d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-16000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 16001-17000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 17001-18000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 18001-19000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 19001-20000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 20001-21000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 21001-22000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 22001-23000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 23001-24000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 24001-25000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 25001-26000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 26001-27000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 27001-28000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 28001-29000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 29001-30000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 30001-31000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 31001-32000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 32001-33000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 33001-34000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 34001-35000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 35001-36000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 36001-37000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 37001-38000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 38001-39000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 39001-40000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 40001-41000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 41001-42000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 42001-43000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 43001-44000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 44001-45000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 45001-46000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 46001-47000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 47001-48000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 48001-49000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 49001-50000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 50001-51000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 51001-52000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 52001-53000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 53001-54000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 54001-55000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 55001-56000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 56001-57000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 57001-58000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 58001-59000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 59001-60000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 60001-61000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 61001-62000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 62001-63000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 63001-64000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 64001-65000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 65001-66000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 66001-67000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 67001-68000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 68001-69000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 69001-70000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 70001-70446...\n",
      "INFO:biothings.client:done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000234956</th>\n",
       "      <td>100507406</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 2539</td>\n",
       "      <td>LINC02539</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000260584</th>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000100376</th>\n",
       "      <td>55007</td>\n",
       "      <td>3.0</td>\n",
       "      <td>family with sequence similarity 118 member A</td>\n",
       "      <td>FAM118A</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>C22orf8</td>\n",
       "      <td>Enables identical protein binding activity. Pr...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215458</th>\n",
       "      <td>284837</td>\n",
       "      <td>2.0</td>\n",
       "      <td>apoptosis associated transcript in bladder cancer</td>\n",
       "      <td>AATBC</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000236028</th>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             _id  _version  \\\n",
       "ensembl_gene_id                              \n",
       "ENSG00000234956        100507406       1.0   \n",
       "ENSG00000260584  ENSG00000260584       1.0   \n",
       "ENSG00000100376            55007       3.0   \n",
       "ENSG00000215458           284837       2.0   \n",
       "ENSG00000236028  ENSG00000236028       1.0   \n",
       "\n",
       "                                                              name     symbol  \\\n",
       "ensembl_gene_id                                                                 \n",
       "ENSG00000234956        long intergenic non-protein coding RNA 2539  LINC02539   \n",
       "ENSG00000260584                                                NaN        NaN   \n",
       "ENSG00000100376       family with sequence similarity 118 member A    FAM118A   \n",
       "ENSG00000215458  apoptosis associated transcript in bladder cancer      AATBC   \n",
       "ENSG00000236028                                                NaN        NaN   \n",
       "\n",
       "                   type_of_gene    alias  \\\n",
       "ensembl_gene_id                            \n",
       "ENSG00000234956           ncRNA      NaN   \n",
       "ENSG00000260584             NaN      NaN   \n",
       "ENSG00000100376  protein-coding  C22orf8   \n",
       "ENSG00000215458           ncRNA      NaN   \n",
       "ENSG00000236028             NaN      NaN   \n",
       "\n",
       "                                                           summary notfound  \n",
       "ensembl_gene_id                                                              \n",
       "ENSG00000234956                                                NaN      NaN  \n",
       "ENSG00000260584                                                NaN      NaN  \n",
       "ENSG00000100376  Enables identical protein binding activity. Pr...      NaN  \n",
       "ENSG00000215458                                                NaN      NaN  \n",
       "ENSG00000236028                                                NaN      NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "mygene_output = mg.getgenes(ensembl_ids_df['ensembl_gene_id'], \n",
    "                            fields=[\"symbol\", \"name\", \"summary\", \"type_of_gene\", \"alias\"], \n",
    "                            as_dataframe=True)\n",
    "\n",
    "mygene_output.index.rename(\"ensembl_gene_id\", inplace=True)\n",
    "mygene_output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23bb114e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotations found for 70106 genes.\n",
      "No annotations found for 1152 genes.\n"
     ]
    }
   ],
   "source": [
    "print(\"Annotations found for \" + str(sum(mygene_output['notfound'].isna())) + \" genes.\")\n",
    "print(\"No annotations found for \" + str(sum(mygene_output['notfound'] == True)) + \" genes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf71470",
   "metadata": {},
   "source": [
    "# Part 2: Clean the data\n",
    "\n",
    "## Join and standardize columns / values\n",
    "\n",
    "For consistency with the `agora-data-tools` transform process, this uses the etl standardize functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "186d8cb8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(71258, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000234956</td>\n",
       "      <td>100507406</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 2539</td>\n",
       "      <td>LINC02539</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100376</td>\n",
       "      <td>55007</td>\n",
       "      <td>3.0</td>\n",
       "      <td>family with sequence similarity 118 member A</td>\n",
       "      <td>FAM118A</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>C22orf8</td>\n",
       "      <td>Enables identical protein binding activity. Pr...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000215458</td>\n",
       "      <td>284837</td>\n",
       "      <td>2.0</td>\n",
       "      <td>apoptosis associated transcript in bladder cancer</td>\n",
       "      <td>AATBC</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id              _id  _version  \\\n",
       "0  ENSG00000234956        100507406       1.0   \n",
       "1  ENSG00000260584  ENSG00000260584       1.0   \n",
       "2  ENSG00000100376            55007       3.0   \n",
       "3  ENSG00000215458           284837       2.0   \n",
       "4  ENSG00000236028  ENSG00000236028       1.0   \n",
       "\n",
       "                                                name     symbol  \\\n",
       "0        long intergenic non-protein coding RNA 2539  LINC02539   \n",
       "1                                                NaN        NaN   \n",
       "2       family with sequence similarity 118 member A    FAM118A   \n",
       "3  apoptosis associated transcript in bladder cancer      AATBC   \n",
       "4                                                NaN        NaN   \n",
       "\n",
       "     type_of_gene    alias                                            summary  \\\n",
       "0           ncRNA      NaN                                                NaN   \n",
       "1             NaN      NaN                                                NaN   \n",
       "2  protein-coding  C22orf8  Enables identical protein binding activity. Pr...   \n",
       "3           ncRNA      NaN                                                NaN   \n",
       "4             NaN      NaN                                                NaN   \n",
       "\n",
       "  notfound  \n",
       "0      NaN  \n",
       "1      NaN  \n",
       "2      NaN  \n",
       "3      NaN  \n",
       "4      NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This merge may not be strictly necessary? mygene should return at least one row all genes queried even if  \n",
    "# it can't find the gene in the database\n",
    "gene_table_merged = pd.merge(left = ensembl_ids_df, right = mygene_output, how = 'left', on = 'ensembl_gene_id')\n",
    "\n",
    "gene_table_merged = transform.standardize_column_names(gene_table_merged)\n",
    "gene_table_merged = transform.standardize_values(gene_table_merged)\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9441115e",
   "metadata": {},
   "source": [
    "## Fix alias field\n",
    "\n",
    "Fix `NaN` values in the `alias` field and make sure every alias value is a list, not a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "285c10d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NaN or NULL alias values become empty lists\n",
    "for row in gene_table_merged.loc[gene_table_merged['alias'].isnull(), 'alias'].index:\n",
    "    gene_table_merged.at[row, 'alias'] = []\n",
    "\n",
    "# Some alias values are a single string, not a list. Turn them into lists here.\n",
    "gene_table_merged['alias'] = gene_table_merged['alias'].apply(lambda cell: cell if isinstance(cell, list) else [cell])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0be1b92",
   "metadata": {},
   "source": [
    "## Remove duplicate Ensembl IDs from the list. \n",
    "\n",
    "Duplicates in the list typically have the same Ensembl ID but different gene symbols. This usually happens when a single Ensembl ID maps to multiple Entrez IDs in the NCBI database. There's not a good way to reconcile this, so just use the first entry in the list for each ensembl ID and discard the rest, which is what the Agora front end does. The gene symbols of duplicate rows are then added as aliases to the matching unique row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bc63cc53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124905411</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905411</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124906452</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906452</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124906454</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906454</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124906457</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906457</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57793</th>\n",
       "      <td>ENSG00000273730</td>\n",
       "      <td>124908522</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124908522</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63658</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63659</th>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>285626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC285626</td>\n",
       "      <td>LOC285626</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70770</th>\n",
       "      <td>ENSG00000260788</td>\n",
       "      <td>105371366</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105371366</td>\n",
       "      <td>LOC105371366</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70771</th>\n",
       "      <td>ENSG00000260788</td>\n",
       "      <td>124903732</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CDH13 antisense oligodendrocyte and neuron ass...</td>\n",
       "      <td>CEDORA</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[FEDORA]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>861 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id        _id  _version  \\\n",
       "127    ENSG00000284156  124905408       2.0   \n",
       "128    ENSG00000284156  124905411       2.0   \n",
       "129    ENSG00000284156  124906452       2.0   \n",
       "130    ENSG00000284156  124906454       2.0   \n",
       "131    ENSG00000284156  124906457       2.0   \n",
       "...                ...        ...       ...   \n",
       "57793  ENSG00000273730  124908522       1.0   \n",
       "63658  ENSG00000249738  105377683       1.0   \n",
       "63659  ENSG00000249738     285626       1.0   \n",
       "70770  ENSG00000260788  105371366       1.0   \n",
       "70771  ENSG00000260788  124903732       2.0   \n",
       "\n",
       "                                                    name        symbol  \\\n",
       "127                            double homeobox protein 4  LOC124905408   \n",
       "128                            double homeobox protein 4  LOC124905411   \n",
       "129                            double homeobox protein 4  LOC124906452   \n",
       "130                            double homeobox protein 4  LOC124906454   \n",
       "131                            double homeobox protein 4  LOC124906457   \n",
       "...                                                  ...           ...   \n",
       "57793                                 5.8S ribosomal RNA  LOC124908522   \n",
       "63658                       uncharacterized LOC105377683  LOC105377683   \n",
       "63659                          uncharacterized LOC285626     LOC285626   \n",
       "70770                       uncharacterized LOC105371366  LOC105371366   \n",
       "70771  CDH13 antisense oligodendrocyte and neuron ass...        CEDORA   \n",
       "\n",
       "         type_of_gene     alias summary notfound  \n",
       "127    protein-coding        []     NaN      NaN  \n",
       "128    protein-coding        []     NaN      NaN  \n",
       "129    protein-coding        []     NaN      NaN  \n",
       "130    protein-coding        []     NaN      NaN  \n",
       "131    protein-coding        []     NaN      NaN  \n",
       "...               ...       ...     ...      ...  \n",
       "57793            rRNA        []     NaN      NaN  \n",
       "63658           ncRNA        []     NaN      NaN  \n",
       "63659           ncRNA        []     NaN      NaN  \n",
       "70770           ncRNA        []     NaN      NaN  \n",
       "70771           ncRNA  [FEDORA]     NaN      NaN  \n",
       "\n",
       "[861 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated() will return true if the ID is a duplicate and is not the first one to appear the list. \n",
    "dupes = gene_table_merged['ensembl_gene_id'].duplicated()\n",
    "dupe_vals = gene_table_merged[dupes]\n",
    "\n",
    "# Rows with duplicated Ensembl IDs\n",
    "gene_table_merged.loc[gene_table_merged['ensembl_gene_id'].isin(dupe_vals['ensembl_gene_id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e69f2c8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(70446, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>127</td>\n",
       "      <td>ENSG00000284156</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906466, LOC124906454, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1343</th>\n",
       "      <td>1360</td>\n",
       "      <td>ENSG00000283878</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2039</th>\n",
       "      <td>2073</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>124905699</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FAM231A/C-like protein LOC102723383</td>\n",
       "      <td>LOC124905699</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905693, LOC124905556, LOC124903857]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4943</th>\n",
       "      <td>4980</td>\n",
       "      <td>ENSG00000278747</td>\n",
       "      <td>124904706</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>LOC124904706</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>[LOC124906683, LOC124904108]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8611</th>\n",
       "      <td>8650</td>\n",
       "      <td>ENSG00000273624</td>\n",
       "      <td>124905327</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>LOC124905327</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>[LOC124905519, LOC124900632]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9582</th>\n",
       "      <td>9623</td>\n",
       "      <td>ENSG00000282767</td>\n",
       "      <td>124900566</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900566</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900359]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15070</th>\n",
       "      <td>15112</td>\n",
       "      <td>ENSG00000284502</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906466, LOC124906454, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15558</th>\n",
       "      <td>15617</td>\n",
       "      <td>ENSG00000280619</td>\n",
       "      <td>107987420</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC107987420</td>\n",
       "      <td>LOC107987420</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC105374682]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16366</th>\n",
       "      <td>16426</td>\n",
       "      <td>ENSG00000283898</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17629</th>\n",
       "      <td>17706</td>\n",
       "      <td>ENSG00000288164</td>\n",
       "      <td>102723573</td>\n",
       "      <td>2.0</td>\n",
       "      <td>40S ribosomal protein S8-like</td>\n",
       "      <td>LOC102723573</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC102724737]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17633</th>\n",
       "      <td>17711</td>\n",
       "      <td>ENSG00000278903</td>\n",
       "      <td>124905527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124905527</td>\n",
       "      <td>LOC124905527</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC124905468, LOC124905312]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19082</th>\n",
       "      <td>19162</td>\n",
       "      <td>ENSG00000284025</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19820</th>\n",
       "      <td>19917</td>\n",
       "      <td>ENSG00000284475</td>\n",
       "      <td>124906466</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906466</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906454, LOC124906463, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20319</th>\n",
       "      <td>20433</td>\n",
       "      <td>ENSG00000267635</td>\n",
       "      <td>100505724</td>\n",
       "      <td>1.0</td>\n",
       "      <td>keratin associated protein 9-7</td>\n",
       "      <td>KRTAP9-7</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[KRTAP9-6, KAP9.7, KRTAP9L1]</td>\n",
       "      <td>Predicted to be located in cytosol. [provided ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22614</th>\n",
       "      <td>22729</td>\n",
       "      <td>ENSG00000276297</td>\n",
       "      <td>107987420</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC107987420</td>\n",
       "      <td>LOC107987420</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC105374682]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22918</th>\n",
       "      <td>23034</td>\n",
       "      <td>ENSG00000283884</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23033</th>\n",
       "      <td>23166</td>\n",
       "      <td>ENSG00000283893</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23458</th>\n",
       "      <td>23608</td>\n",
       "      <td>ENSG00000288249</td>\n",
       "      <td>124907358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124907358</td>\n",
       "      <td>LOC124907358</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124907096]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23545</th>\n",
       "      <td>23696</td>\n",
       "      <td>ENSG00000283830</td>\n",
       "      <td>124906466</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906466</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906454, LOC124906463, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23934</th>\n",
       "      <td>24102</td>\n",
       "      <td>ENSG00000282177</td>\n",
       "      <td>124900357</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900357</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900358, LOC124900359, LOC124900356]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29465</th>\n",
       "      <td>29636</td>\n",
       "      <td>ENSG00000276367</td>\n",
       "      <td>124900356</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900356</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900358, LOC124900359]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29578</th>\n",
       "      <td>29751</td>\n",
       "      <td>ENSG00000284181</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906466, LOC124906454, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31897</th>\n",
       "      <td>32087</td>\n",
       "      <td>ENSG00000275757</td>\n",
       "      <td>124908527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124908527</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[LOC124907119, LOC124907594, LOC124907333, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32924</th>\n",
       "      <td>33320</td>\n",
       "      <td>ENSG00000288184</td>\n",
       "      <td>124907358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC124907358</td>\n",
       "      <td>LOC124907358</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124907096]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34130</th>\n",
       "      <td>34527</td>\n",
       "      <td>ENSG00000276387</td>\n",
       "      <td>124900571</td>\n",
       "      <td>2.0</td>\n",
       "      <td>killer cell immunoglobulin-like receptor 2DS1</td>\n",
       "      <td>LOC124900571</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[KIR2DL1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36185</th>\n",
       "      <td>36583</td>\n",
       "      <td>ENSG00000283820</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124905409, LOC124906466, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37941</th>\n",
       "      <td>38356</td>\n",
       "      <td>ENSG00000284002</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40725</th>\n",
       "      <td>41157</td>\n",
       "      <td>ENSG00000283767</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016</th>\n",
       "      <td>41465</td>\n",
       "      <td>ENSG00000284507</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43177</th>\n",
       "      <td>43643</td>\n",
       "      <td>ENSG00000284383</td>\n",
       "      <td>124906466</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906466</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906454, LOC124906463, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43910</th>\n",
       "      <td>44393</td>\n",
       "      <td>ENSG00000283955</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906466, LOC124906454, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44998</th>\n",
       "      <td>45498</td>\n",
       "      <td>ENSG00000284262</td>\n",
       "      <td>124906466</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906466</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906454, LOC124906463, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45876</th>\n",
       "      <td>46393</td>\n",
       "      <td>ENSG00000275583</td>\n",
       "      <td>124900575</td>\n",
       "      <td>2.0</td>\n",
       "      <td>killer cell immunoglobulin-like receptor 2DS2</td>\n",
       "      <td>LOC124900575</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[KIR2DS2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47128</th>\n",
       "      <td>47646</td>\n",
       "      <td>ENSG00000275921</td>\n",
       "      <td>3806</td>\n",
       "      <td>5.0</td>\n",
       "      <td>killer cell immunoglobulin like receptor, two ...</td>\n",
       "      <td>KIR2DS1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124900571, CD158a, p50.1, CD158H]</td>\n",
       "      <td>Killer cell immunoglobulin-like receptors (KIR...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47396</th>\n",
       "      <td>47915</td>\n",
       "      <td>ENSG00000274175</td>\n",
       "      <td>102724101</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TP53 target 3 family member E</td>\n",
       "      <td>TP53TG3E</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[TP53TG3B, TP53TG3, TP53TG3F, TP53TG3C, TP53TG...</td>\n",
       "      <td>Located in cytoplasm and nucleus. [provided by...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47815</th>\n",
       "      <td>48336</td>\n",
       "      <td>ENSG00000275648</td>\n",
       "      <td>124900357</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900357</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900358, LOC124900356]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48136</th>\n",
       "      <td>48659</td>\n",
       "      <td>ENSG00000278294</td>\n",
       "      <td>124908250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124908250</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[LOC124907485, LOC124907156]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50013</th>\n",
       "      <td>50538</td>\n",
       "      <td>ENSG00000284066</td>\n",
       "      <td>124906466</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124906466</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906454, LOC124906463, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51175</th>\n",
       "      <td>51717</td>\n",
       "      <td>ENSG00000275394</td>\n",
       "      <td>124900357</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U8 small nucleolar RNA</td>\n",
       "      <td>LOC124900357</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>[LOC124900358, LOC124900359, LOC124900356]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52359</th>\n",
       "      <td>52904</td>\n",
       "      <td>ENSG00000284210</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124905409, LOC124906466, LOC124906454, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53208</th>\n",
       "      <td>53770</td>\n",
       "      <td>ENSG00000284174</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124906460, LOC124905409, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55060</th>\n",
       "      <td>55639</td>\n",
       "      <td>ENSG00000275809</td>\n",
       "      <td>124905326</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>LOC124905326</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>[LOC124905327, LOC124900632, LOC124905519]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55372</th>\n",
       "      <td>55954</td>\n",
       "      <td>ENSG00000277660</td>\n",
       "      <td>124904706</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>LOC124904706</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>[LOC124906683, LOC124904108]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55823</th>\n",
       "      <td>56407</td>\n",
       "      <td>ENSG00000229425</td>\n",
       "      <td>105369302</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105369302</td>\n",
       "      <td>LOC105369302</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC101927745]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56927</th>\n",
       "      <td>57512</td>\n",
       "      <td>ENSG00000284496</td>\n",
       "      <td>124905408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>double homeobox protein 4</td>\n",
       "      <td>LOC124905408</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[LOC124906458, LOC124905409, LOC124906466, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56967</th>\n",
       "      <td>57569</td>\n",
       "      <td>ENSG00000288205</td>\n",
       "      <td>124904706</td>\n",
       "      <td>1.0</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>LOC124904706</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>[LOC124906683, LOC124904108]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56983</th>\n",
       "      <td>57587</td>\n",
       "      <td>ENSG00000273730</td>\n",
       "      <td>124908527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8S ribosomal RNA</td>\n",
       "      <td>LOC124908527</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>[LOC124907119, LOC124907594, LOC124907333, LOC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62848</th>\n",
       "      <td>63658</td>\n",
       "      <td>ENSG00000249738</td>\n",
       "      <td>105377683</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105377683</td>\n",
       "      <td>LOC105377683</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[LOC285626]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69959</th>\n",
       "      <td>70770</td>\n",
       "      <td>ENSG00000260788</td>\n",
       "      <td>105371366</td>\n",
       "      <td>1.0</td>\n",
       "      <td>uncharacterized LOC105371366</td>\n",
       "      <td>LOC105371366</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[CEDORA]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       index  ensembl_gene_id        _id  _version  \\\n",
       "127      127  ENSG00000284156  124905408       2.0   \n",
       "1343    1360  ENSG00000283878  124905408       2.0   \n",
       "2039    2073  ENSG00000268674  124905699       1.0   \n",
       "4943    4980  ENSG00000278747  124904706       1.0   \n",
       "8611    8650  ENSG00000273624  124905327       1.0   \n",
       "9582    9623  ENSG00000282767  124900566       1.0   \n",
       "15070  15112  ENSG00000284502  124905408       2.0   \n",
       "15558  15617  ENSG00000280619  107987420       1.0   \n",
       "16366  16426  ENSG00000283898  124905408       2.0   \n",
       "17629  17706  ENSG00000288164  102723573       2.0   \n",
       "17633  17711  ENSG00000278903  124905527       1.0   \n",
       "19082  19162  ENSG00000284025  124905408       2.0   \n",
       "19820  19917  ENSG00000284475  124906466       2.0   \n",
       "20319  20433  ENSG00000267635  100505724       1.0   \n",
       "22614  22729  ENSG00000276297  107987420       1.0   \n",
       "22918  23034  ENSG00000283884  124905408       2.0   \n",
       "23033  23166  ENSG00000283893  124905408       2.0   \n",
       "23458  23608  ENSG00000288249  124907358       1.0   \n",
       "23545  23696  ENSG00000283830  124906466       2.0   \n",
       "23934  24102  ENSG00000282177  124900357       1.0   \n",
       "29465  29636  ENSG00000276367  124900356       1.0   \n",
       "29578  29751  ENSG00000284181  124905408       2.0   \n",
       "31897  32087  ENSG00000275757  124908527       1.0   \n",
       "32924  33320  ENSG00000288184  124907358       1.0   \n",
       "34130  34527  ENSG00000276387  124900571       2.0   \n",
       "36185  36583  ENSG00000283820  124905408       2.0   \n",
       "37941  38356  ENSG00000284002  124905408       2.0   \n",
       "40725  41157  ENSG00000283767  124905408       2.0   \n",
       "41016  41465  ENSG00000284507  124905408       2.0   \n",
       "43177  43643  ENSG00000284383  124906466       2.0   \n",
       "43910  44393  ENSG00000283955  124905408       2.0   \n",
       "44998  45498  ENSG00000284262  124906466       2.0   \n",
       "45876  46393  ENSG00000275583  124900575       2.0   \n",
       "47128  47646  ENSG00000275921       3806       5.0   \n",
       "47396  47915  ENSG00000274175  102724101       1.0   \n",
       "47815  48336  ENSG00000275648  124900357       1.0   \n",
       "48136  48659  ENSG00000278294  124908250       1.0   \n",
       "50013  50538  ENSG00000284066  124906466       2.0   \n",
       "51175  51717  ENSG00000275394  124900357       1.0   \n",
       "52359  52904  ENSG00000284210  124905408       2.0   \n",
       "53208  53770  ENSG00000284174  124905408       2.0   \n",
       "55060  55639  ENSG00000275809  124905326       1.0   \n",
       "55372  55954  ENSG00000277660  124904706       1.0   \n",
       "55823  56407  ENSG00000229425  105369302       1.0   \n",
       "56927  57512  ENSG00000284496  124905408       2.0   \n",
       "56967  57569  ENSG00000288205  124904706       1.0   \n",
       "56983  57587  ENSG00000273730  124908527       1.0   \n",
       "62848  63658  ENSG00000249738  105377683       1.0   \n",
       "69959  70770  ENSG00000260788  105371366       1.0   \n",
       "\n",
       "                                                    name        symbol  \\\n",
       "127                            double homeobox protein 4  LOC124905408   \n",
       "1343                           double homeobox protein 4  LOC124905408   \n",
       "2039                 FAM231A/C-like protein LOC102723383  LOC124905699   \n",
       "4943                                 U6 spliceosomal RNA  LOC124904706   \n",
       "8611                                 U6 spliceosomal RNA  LOC124905327   \n",
       "9582                              U8 small nucleolar RNA  LOC124900566   \n",
       "15070                          double homeobox protein 4  LOC124905408   \n",
       "15558                       uncharacterized LOC107987420  LOC107987420   \n",
       "16366                          double homeobox protein 4  LOC124905408   \n",
       "17629                      40S ribosomal protein S8-like  LOC102723573   \n",
       "17633                       uncharacterized LOC124905527  LOC124905527   \n",
       "19082                          double homeobox protein 4  LOC124905408   \n",
       "19820                          double homeobox protein 4  LOC124906466   \n",
       "20319                     keratin associated protein 9-7      KRTAP9-7   \n",
       "22614                       uncharacterized LOC107987420  LOC107987420   \n",
       "22918                          double homeobox protein 4  LOC124905408   \n",
       "23033                          double homeobox protein 4  LOC124905408   \n",
       "23458                       uncharacterized LOC124907358  LOC124907358   \n",
       "23545                          double homeobox protein 4  LOC124906466   \n",
       "23934                             U8 small nucleolar RNA  LOC124900357   \n",
       "29465                             U8 small nucleolar RNA  LOC124900356   \n",
       "29578                          double homeobox protein 4  LOC124905408   \n",
       "31897                                 5.8S ribosomal RNA  LOC124908527   \n",
       "32924                       uncharacterized LOC124907358  LOC124907358   \n",
       "34130      killer cell immunoglobulin-like receptor 2DS1  LOC124900571   \n",
       "36185                          double homeobox protein 4  LOC124905408   \n",
       "37941                          double homeobox protein 4  LOC124905408   \n",
       "40725                          double homeobox protein 4  LOC124905408   \n",
       "41016                          double homeobox protein 4  LOC124905408   \n",
       "43177                          double homeobox protein 4  LOC124906466   \n",
       "43910                          double homeobox protein 4  LOC124905408   \n",
       "44998                          double homeobox protein 4  LOC124906466   \n",
       "45876      killer cell immunoglobulin-like receptor 2DS2  LOC124900575   \n",
       "47128  killer cell immunoglobulin like receptor, two ...       KIR2DS1   \n",
       "47396                      TP53 target 3 family member E      TP53TG3E   \n",
       "47815                             U8 small nucleolar RNA  LOC124900357   \n",
       "48136                                 5.8S ribosomal RNA  LOC124908250   \n",
       "50013                          double homeobox protein 4  LOC124906466   \n",
       "51175                             U8 small nucleolar RNA  LOC124900357   \n",
       "52359                          double homeobox protein 4  LOC124905408   \n",
       "53208                          double homeobox protein 4  LOC124905408   \n",
       "55060                                U6 spliceosomal RNA  LOC124905326   \n",
       "55372                                U6 spliceosomal RNA  LOC124904706   \n",
       "55823                       uncharacterized LOC105369302  LOC105369302   \n",
       "56927                          double homeobox protein 4  LOC124905408   \n",
       "56967                                U6 spliceosomal RNA  LOC124904706   \n",
       "56983                                 5.8S ribosomal RNA  LOC124908527   \n",
       "62848                       uncharacterized LOC105377683  LOC105377683   \n",
       "69959                       uncharacterized LOC105371366  LOC105371366   \n",
       "\n",
       "         type_of_gene                                              alias  \\\n",
       "127    protein-coding  [LOC124905409, LOC124906466, LOC124906454, LOC...   \n",
       "1343   protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "2039   protein-coding         [LOC124905693, LOC124905556, LOC124903857]   \n",
       "4943            snRNA                       [LOC124906683, LOC124904108]   \n",
       "8611            snRNA                       [LOC124905519, LOC124900632]   \n",
       "9582           snoRNA                                     [LOC124900359]   \n",
       "15070  protein-coding  [LOC124905409, LOC124906466, LOC124906454, LOC...   \n",
       "15558           ncRNA                                     [LOC105374682]   \n",
       "16366  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "17629  protein-coding                                     [LOC102724737]   \n",
       "17633           ncRNA                       [LOC124905468, LOC124905312]   \n",
       "19082  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "19820  protein-coding  [LOC124905409, LOC124906454, LOC124906463, LOC...   \n",
       "20319  protein-coding                       [KRTAP9-6, KAP9.7, KRTAP9L1]   \n",
       "22614           ncRNA                                     [LOC105374682]   \n",
       "22918  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "23033  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "23458  protein-coding                                     [LOC124907096]   \n",
       "23545  protein-coding  [LOC124905409, LOC124906454, LOC124906463, LOC...   \n",
       "23934          snoRNA         [LOC124900358, LOC124900359, LOC124900356]   \n",
       "29465          snoRNA                       [LOC124900358, LOC124900359]   \n",
       "29578  protein-coding  [LOC124905409, LOC124906466, LOC124906454, LOC...   \n",
       "31897            rRNA  [LOC124907119, LOC124907594, LOC124907333, LOC...   \n",
       "32924  protein-coding                                     [LOC124907096]   \n",
       "34130  protein-coding                                          [KIR2DL1]   \n",
       "36185  protein-coding  [LOC124906458, LOC124905409, LOC124906466, LOC...   \n",
       "37941  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "40725  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "41016  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "43177  protein-coding  [LOC124905409, LOC124906454, LOC124906463, LOC...   \n",
       "43910  protein-coding  [LOC124905409, LOC124906466, LOC124906454, LOC...   \n",
       "44998  protein-coding  [LOC124905409, LOC124906454, LOC124906463, LOC...   \n",
       "45876  protein-coding                                          [KIR2DS2]   \n",
       "47128  protein-coding              [LOC124900571, CD158a, p50.1, CD158H]   \n",
       "47396  protein-coding  [TP53TG3B, TP53TG3, TP53TG3F, TP53TG3C, TP53TG...   \n",
       "47815          snoRNA                       [LOC124900358, LOC124900356]   \n",
       "48136            rRNA                       [LOC124907485, LOC124907156]   \n",
       "50013  protein-coding  [LOC124905409, LOC124906454, LOC124906463, LOC...   \n",
       "51175          snoRNA         [LOC124900358, LOC124900359, LOC124900356]   \n",
       "52359  protein-coding  [LOC124905409, LOC124906466, LOC124906454, LOC...   \n",
       "53208  protein-coding  [LOC124906458, LOC124906460, LOC124905409, LOC...   \n",
       "55060           snRNA         [LOC124905327, LOC124900632, LOC124905519]   \n",
       "55372           snRNA                       [LOC124906683, LOC124904108]   \n",
       "55823           ncRNA                                     [LOC101927745]   \n",
       "56927  protein-coding  [LOC124906458, LOC124905409, LOC124906466, LOC...   \n",
       "56967           snRNA                       [LOC124906683, LOC124904108]   \n",
       "56983            rRNA  [LOC124907119, LOC124907594, LOC124907333, LOC...   \n",
       "62848           ncRNA                                        [LOC285626]   \n",
       "69959           ncRNA                                           [CEDORA]   \n",
       "\n",
       "                                                 summary notfound  \n",
       "127                                                  NaN      NaN  \n",
       "1343                                                 NaN      NaN  \n",
       "2039                                                 NaN      NaN  \n",
       "4943                                                 NaN      NaN  \n",
       "8611                                                 NaN      NaN  \n",
       "9582                                                 NaN      NaN  \n",
       "15070                                                NaN      NaN  \n",
       "15558                                                NaN      NaN  \n",
       "16366                                                NaN      NaN  \n",
       "17629                                                NaN      NaN  \n",
       "17633                                                NaN      NaN  \n",
       "19082                                                NaN      NaN  \n",
       "19820                                                NaN      NaN  \n",
       "20319  Predicted to be located in cytosol. [provided ...      NaN  \n",
       "22614                                                NaN      NaN  \n",
       "22918                                                NaN      NaN  \n",
       "23033                                                NaN      NaN  \n",
       "23458                                                NaN      NaN  \n",
       "23545                                                NaN      NaN  \n",
       "23934                                                NaN      NaN  \n",
       "29465                                                NaN      NaN  \n",
       "29578                                                NaN      NaN  \n",
       "31897                                                NaN      NaN  \n",
       "32924                                                NaN      NaN  \n",
       "34130                                                NaN      NaN  \n",
       "36185                                                NaN      NaN  \n",
       "37941                                                NaN      NaN  \n",
       "40725                                                NaN      NaN  \n",
       "41016                                                NaN      NaN  \n",
       "43177                                                NaN      NaN  \n",
       "43910                                                NaN      NaN  \n",
       "44998                                                NaN      NaN  \n",
       "45876                                                NaN      NaN  \n",
       "47128  Killer cell immunoglobulin-like receptors (KIR...      NaN  \n",
       "47396  Located in cytoplasm and nucleus. [provided by...      NaN  \n",
       "47815                                                NaN      NaN  \n",
       "48136                                                NaN      NaN  \n",
       "50013                                                NaN      NaN  \n",
       "51175                                                NaN      NaN  \n",
       "52359                                                NaN      NaN  \n",
       "53208                                                NaN      NaN  \n",
       "55060                                                NaN      NaN  \n",
       "55372                                                NaN      NaN  \n",
       "55823                                                NaN      NaN  \n",
       "56927                                                NaN      NaN  \n",
       "56967                                                NaN      NaN  \n",
       "56983                                                NaN      NaN  \n",
       "62848                                                NaN      NaN  \n",
       "69959                                                NaN      NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove duplicates from the list\n",
    "gene_table_merged = gene_table_merged[dupes == False].reset_index()\n",
    "\n",
    "# For each duplicate row, add its symbol as an alias\n",
    "for row in dupe_vals.index:\n",
    "    match = gene_table_merged['ensembl_gene_id'] == dupe_vals['ensembl_gene_id'][row]\n",
    "    match_ind = gene_table_merged[match].index[0] # There should only be one row\n",
    "\n",
    "    # Add the duplicate's symbol to the alias list\n",
    "    gene_table_merged.at[match_ind, 'alias'].append(dupe_vals['symbol'][row])\n",
    "    \n",
    "    # Make sure we didn't add duplicate aliases\n",
    "    gene_table_merged.at[match_ind, 'alias'] = list(set(gene_table_merged.at[match_ind, 'alias']))\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "\n",
    "# Printed out table should have unique Ensembl IDs with aliases properly added\n",
    "gene_table_merged.loc[gene_table_merged['ensembl_gene_id'].isin(dupe_vals['ensembl_gene_id'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "702c361d",
   "metadata": {},
   "source": [
    "# Part 3: Create Ensembl archive permalinks\n",
    "\n",
    "## Get a table of Ensembl archive URLs\n",
    "\n",
    "This is where we need to use the R biomaRt library specifically, instead of any of the available Python interfaces to Biomart, to get a table of Ensembl release versions and their corresponding archive URLs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4a1bbdee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             name     date                                 url version\n",
      "1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n",
      "2     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n",
      "3     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n",
      "4     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n",
      "5     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n",
      "6     Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n",
      "7     Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n",
      "8     Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n",
      "9     Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n",
      "10    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n",
      "11    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n",
      "12     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n",
      "13     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n",
      "14     Ensembl 97 Jul 2019 https://jul2019.archive.ensembl.org      97\n",
      "15     Ensembl 96 Apr 2019 https://apr2019.archive.ensembl.org      96\n",
      "16     Ensembl 95 Jan 2019 https://jan2019.archive.ensembl.org      95\n",
      "17     Ensembl 94 Oct 2018 https://oct2018.archive.ensembl.org      94\n",
      "18     Ensembl 93 Jul 2018 https://jul2018.archive.ensembl.org      93\n",
      "19     Ensembl 92 Apr 2018 https://apr2018.archive.ensembl.org      92\n",
      "20     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80\n",
      "21     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n",
      "22     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n",
      "23     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n",
      "   current_release\n",
      "1                 \n",
      "2                *\n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10                \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n",
      "17                \n",
      "18                \n",
      "19                \n",
      "20                \n",
      "21                \n",
      "22                \n",
      "23                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "archive_df = r.listEnsemblArchives()\n",
    "archive_df.to_csvfile(path = archive_filename, row_names = False, quote = False)\n",
    "\n",
    "print(archive_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e150d8e",
   "metadata": {},
   "source": [
    "## Query Ensembl for each gene's version\n",
    "\n",
    "Ensembl's REST API can only take 1000 genes at once, so this is looped to query groups of 1000. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9a747309",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70446\n",
      "Querying genes 1 - 1000\n",
      "Querying genes 1001 - 2000\n",
      "Querying genes 2001 - 3000\n",
      "Querying genes 3001 - 4000\n",
      "Querying genes 4001 - 5000\n",
      "Querying genes 5001 - 6000\n",
      "Querying genes 6001 - 7000\n",
      "Querying genes 7001 - 8000\n",
      "Querying genes 8001 - 9000\n",
      "Querying genes 9001 - 10000\n",
      "Querying genes 10001 - 11000\n",
      "Querying genes 11001 - 12000\n",
      "Querying genes 12001 - 13000\n",
      "Querying genes 13001 - 14000\n",
      "Querying genes 14001 - 15000\n",
      "Querying genes 15001 - 16000\n",
      "Querying genes 16001 - 17000\n",
      "Querying genes 17001 - 18000\n",
      "Querying genes 18001 - 19000\n",
      "Querying genes 19001 - 20000\n",
      "Querying genes 20001 - 21000\n",
      "Querying genes 21001 - 22000\n",
      "Querying genes 22001 - 23000\n",
      "Querying genes 23001 - 24000\n",
      "Querying genes 24001 - 25000\n",
      "Querying genes 25001 - 26000\n",
      "Querying genes 26001 - 27000\n",
      "Error retrieving Ensembl versions for genes 26001 - 27000. Trying again...\n",
      "Error retrieving Ensembl versions for genes 26001 - 27000. Trying again...\n",
      "Querying genes 27001 - 28000\n",
      "Querying genes 28001 - 29000\n",
      "Querying genes 29001 - 30000\n",
      "Error retrieving Ensembl versions for genes 29001 - 30000. Trying again...\n",
      "Error retrieving Ensembl versions for genes 29001 - 30000. Trying again...\n",
      "Querying genes 30001 - 31000\n",
      "Querying genes 31001 - 32000\n",
      "Querying genes 32001 - 33000\n",
      "Querying genes 33001 - 34000\n",
      "Querying genes 34001 - 35000\n",
      "Querying genes 35001 - 36000\n",
      "Querying genes 36001 - 37000\n",
      "Querying genes 37001 - 38000\n",
      "Querying genes 38001 - 39000\n",
      "Querying genes 39001 - 40000\n",
      "Querying genes 40001 - 41000\n",
      "Querying genes 41001 - 42000\n",
      "Querying genes 42001 - 43000\n",
      "Querying genes 43001 - 44000\n",
      "Querying genes 44001 - 45000\n",
      "Querying genes 45001 - 46000\n",
      "Querying genes 46001 - 47000\n",
      "Querying genes 47001 - 48000\n",
      "Querying genes 48001 - 49000\n",
      "Querying genes 49001 - 50000\n",
      "Querying genes 50001 - 51000\n",
      "Querying genes 51001 - 52000\n",
      "Querying genes 52001 - 53000\n",
      "Querying genes 53001 - 54000\n",
      "Querying genes 54001 - 55000\n",
      "Querying genes 55001 - 56000\n",
      "Querying genes 56001 - 57000\n",
      "Querying genes 57001 - 58000\n",
      "Querying genes 58001 - 59000\n",
      "Querying genes 59001 - 60000\n",
      "Querying genes 60001 - 61000\n",
      "Querying genes 61001 - 62000\n",
      "Querying genes 62001 - 63000\n",
      "Querying genes 63001 - 64000\n",
      "Querying genes 64001 - 65000\n",
      "Querying genes 65001 - 66000\n",
      "Querying genes 66001 - 67000\n",
      "Querying genes 67001 - 68000\n",
      "Querying genes 68001 - 69000\n",
      "Error retrieving Ensembl versions for genes 68001 - 69000. Trying again...\n",
      "Error retrieving Ensembl versions for genes 68001 - 69000. Trying again...\n",
      "Querying genes 69001 - 70000\n",
      "Querying genes 70001 - 70446\n",
      "70446\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>version</th>\n",
       "      <th>peptide</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>is_current</th>\n",
       "      <th>release</th>\n",
       "      <th>possible_replacement</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70441</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000233405.2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000233405</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70442</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000284485.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000284485</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70443</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000234640.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000234640</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70444</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000188596.11</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000188596</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70445</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000251205.3</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000251205</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      assembly  version peptide              latest  type               id  \\\n",
       "70441   GRCh38        2    None   ENSG00000233405.2  Gene  ENSG00000233405   \n",
       "70442   GRCh38        1    None   ENSG00000284485.1  Gene  ENSG00000284485   \n",
       "70443   GRCh38        1    None   ENSG00000234640.1  Gene  ENSG00000234640   \n",
       "70444   GRCh38       11    None  ENSG00000188596.11  Gene  ENSG00000188596   \n",
       "70445   GRCh38        3    None   ENSG00000251205.3  Gene  ENSG00000251205   \n",
       "\n",
       "      is_current release possible_replacement  \n",
       "70441          1     109                   []  \n",
       "70442          1     109                   []  \n",
       "70443          1     109                   []  \n",
       "70444          1     109                   []  \n",
       "70445          1     109                   []  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://rest.ensembl.org/archive/id\"\n",
    "headers = {\"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "\n",
    "ids = gene_table_merged['ensembl_gene_id'].tolist()\n",
    "print(len(ids))\n",
    "\n",
    "# We can only query 1000 genes at a time\n",
    "batch_ind = range(0, len(ids), 1000)\n",
    "results = []\n",
    "\n",
    "for B in batch_ind:\n",
    "    end = min(len(ids), B + 1000)\n",
    "    print(\"Querying genes \" + str(B+1) + \" - \" + str(end))\n",
    "    \n",
    "    request_data = '{ \"id\" : ' + str(ids[B:end]) + ' }'\n",
    "    request_data = request_data.replace(\"'\", \"\\\"\")\n",
    "    \n",
    "    ok = False\n",
    "    tries = 0\n",
    "    \n",
    "    while tries < 5 and not ok:\n",
    "        try:\n",
    "            res = requests.post(url, headers=headers, data=request_data)\n",
    "            ok = res.ok\n",
    "        except:\n",
    "            ok = False\n",
    "        \n",
    "        tries = tries + 1\n",
    "        \n",
    "        if not ok:\n",
    "            #res.raise_for_status()\n",
    "            print(\"Error retrieving Ensembl versions for genes \" + str(B+1) + \" - \" + str(end) + \n",
    "                  \". Trying again...\")\n",
    "        else:\n",
    "            results = results + res.json()\n",
    "            break\n",
    "\n",
    "print(len(results))\n",
    "\n",
    "versions = pd.json_normalize(results)\n",
    "\n",
    "versions.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5c108238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release\n",
       "100       23\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       32\n",
       "107       10\n",
       "108        4\n",
       "109    69299\n",
       "80        21\n",
       "81         2\n",
       "82        10\n",
       "84       673\n",
       "87        61\n",
       "89        20\n",
       "91        75\n",
       "93        53\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         8\n",
       "99         6\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions.groupby('release').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bf5aecb1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70446\n",
      "70446\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Check that all IDs are the same between the result and the gene table\n",
    "print(len(versions['id']))\n",
    "print(len(gene_table_merged))\n",
    "print(all(versions['id'].isin(gene_table_merged['ensembl_gene_id'])) and \n",
    "      all(gene_table_merged['ensembl_gene_id'].isin(versions['id'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7fc8bbcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure everything is GRCh38, not GRCh37\n",
    "all(versions['assembly'] == \"GRCh38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e29252",
   "metadata": {},
   "source": [
    "## Create permalinks based on archive version\n",
    "\n",
    "**Not all of these versions have an archive.** We can go back to the closest previous archive for these but the link isn't guaranteed to work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0d5b5652",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "archive_table = pd.read_csv(archive_filename)\n",
    "\n",
    "# Remove GRCh37 from the archive list\n",
    "archive_table = archive_table[archive_table['version'] != \"GRCh37\"].reset_index()\n",
    "\n",
    "archive_table['numeric_version'] = archive_table['version'].astype(int)\n",
    "\n",
    "def closest_release(release, archive_table):\n",
    "    if release in archive_table:\n",
    "        return release\n",
    "    \n",
    "    return max([V for V in archive_table['numeric_version'] if V <= release])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "337b2890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "closest_release\n",
       "80       862\n",
       "93        53\n",
       "95        33\n",
       "96        31\n",
       "97        18\n",
       "98         8\n",
       "99         6\n",
       "100       23\n",
       "101        8\n",
       "102       16\n",
       "103       15\n",
       "104       19\n",
       "105        9\n",
       "106       32\n",
       "107       10\n",
       "108        4\n",
       "109    69299\n",
       "dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions['closest_release'] = 0\n",
    "\n",
    "releases = versions['release'].drop_duplicates().astype(int)\n",
    "\n",
    "# Only have to call closest_release once per version, instead of >70k times\n",
    "for release in releases:\n",
    "    versions.loc[versions['release'] == str(release), 'closest_release'] = closest_release(release, archive_table)\n",
    "    \n",
    "versions.groupby('closest_release').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "343e5006",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>version</th>\n",
       "      <th>peptide</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>is_current</th>\n",
       "      <th>release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000234956.7</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000234956</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>109</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000260584.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>109</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000100376.12</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000100376</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>109</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>8</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000215458.8</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000215458</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>109</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000236028.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>109</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly  version peptide              latest  type               id  \\\n",
       "0   GRCh38        7    None   ENSG00000234956.7  Gene  ENSG00000234956   \n",
       "1   GRCh38        1    None   ENSG00000260584.1  Gene  ENSG00000260584   \n",
       "2   GRCh38       12    None  ENSG00000100376.12  Gene  ENSG00000100376   \n",
       "3   GRCh38        8    None   ENSG00000215458.8  Gene  ENSG00000215458   \n",
       "4   GRCh38        1    None   ENSG00000236028.1  Gene  ENSG00000236028   \n",
       "\n",
       "  is_current release possible_replacement  closest_release  \\\n",
       "0          1     109                   []              109   \n",
       "1          1     109                   []              109   \n",
       "2          1     109                   []              109   \n",
       "3          1     109                   []              109   \n",
       "4          1     109                   []              109   \n",
       "\n",
       "                                           permalink  \n",
       "0  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "1  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "2  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "3  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "4  https://feb2023.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions['permalink'] = ''\n",
    "\n",
    "for i in versions.index:\n",
    "    match = archive_table['numeric_version'] == versions.at[i, 'closest_release']\n",
    "    url = archive_table.loc[match, 'url'].to_string(index = False)\n",
    "    if len(url) > 0:\n",
    "        versions.at[i, 'permalink'] = url + \"/Homo_sapiens/Gene/Summary?db=core;g=\" + versions.at[i, 'id']\n",
    "\n",
    "versions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4b01719d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly</th>\n",
       "      <th>version</th>\n",
       "      <th>peptide</th>\n",
       "      <th>latest</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>is_current</th>\n",
       "      <th>release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>closest_release</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000274983.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000274983</td>\n",
       "      <td></td>\n",
       "      <td>84</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000274155.2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000274155</td>\n",
       "      <td></td>\n",
       "      <td>84</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000207765.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000207765</td>\n",
       "      <td></td>\n",
       "      <td>84</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000276117.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000276117</td>\n",
       "      <td></td>\n",
       "      <td>84</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000278423.1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000278423</td>\n",
       "      <td></td>\n",
       "      <td>84</td>\n",
       "      <td>[]</td>\n",
       "      <td>80</td>\n",
       "      <td>https://may2015.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    assembly  version peptide             latest  type               id  \\\n",
       "225   GRCh38        1    None  ENSG00000274983.1  Gene  ENSG00000274983   \n",
       "389   GRCh38        2    None  ENSG00000274155.2  Gene  ENSG00000274155   \n",
       "424   GRCh38        1    None  ENSG00000207765.1  Gene  ENSG00000207765   \n",
       "569   GRCh38        1    None  ENSG00000276117.1  Gene  ENSG00000276117   \n",
       "818   GRCh38        1    None  ENSG00000278423.1  Gene  ENSG00000278423   \n",
       "\n",
       "    is_current release possible_replacement  closest_release  \\\n",
       "225                 84                   []               80   \n",
       "389                 84                   []               80   \n",
       "424                 84                   []               80   \n",
       "569                 84                   []               80   \n",
       "818                 84                   []               80   \n",
       "\n",
       "                                             permalink  \n",
       "225  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "389  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "424  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "569  https://may2015.archive.ensembl.org/Homo_sapie...  \n",
       "818  https://may2015.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions[versions['closest_release'] < 100].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c4128cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://feb2023.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000234956\n",
      "https://feb2023.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000257530\n"
     ]
    }
   ],
   "source": [
    "print(versions['permalink'][0])\n",
    "print(versions['permalink'][25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "73791e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Does every gene have an associated URL?\n",
    "url_base_len = len(archive_table['url'][0]) + 1\n",
    "all([len(url) > url_base_len for url in versions['permalink']])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b98b62a",
   "metadata": {},
   "source": [
    "# Part 4: Add permalinks to the gene table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f3edfd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(70446, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>_id</th>\n",
       "      <th>_version</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>alias</th>\n",
       "      <th>summary</th>\n",
       "      <th>notfound</th>\n",
       "      <th>ensembl_release</th>\n",
       "      <th>possible_replacement</th>\n",
       "      <th>permalink</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000234956</td>\n",
       "      <td>100507406</td>\n",
       "      <td>1.0</td>\n",
       "      <td>long intergenic non-protein coding RNA 2539</td>\n",
       "      <td>LINC02539</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>ENSG00000260584</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000100376</td>\n",
       "      <td>55007</td>\n",
       "      <td>3.0</td>\n",
       "      <td>family with sequence similarity 118 member A</td>\n",
       "      <td>FAM118A</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>[C22orf8]</td>\n",
       "      <td>Enables identical protein binding activity. Pr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000215458</td>\n",
       "      <td>284837</td>\n",
       "      <td>2.0</td>\n",
       "      <td>apoptosis associated transcript in bladder cancer</td>\n",
       "      <td>AATBC</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>ENSG00000236028</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "      <td>[]</td>\n",
       "      <td>https://feb2023.archive.ensembl.org/Homo_sapie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  ensembl_gene_id              _id  _version  \\\n",
       "0      0  ENSG00000234956        100507406       1.0   \n",
       "1      1  ENSG00000260584  ENSG00000260584       1.0   \n",
       "2      2  ENSG00000100376            55007       3.0   \n",
       "3      3  ENSG00000215458           284837       2.0   \n",
       "4      4  ENSG00000236028  ENSG00000236028       1.0   \n",
       "\n",
       "                                                name     symbol  \\\n",
       "0        long intergenic non-protein coding RNA 2539  LINC02539   \n",
       "1                                                NaN        NaN   \n",
       "2       family with sequence similarity 118 member A    FAM118A   \n",
       "3  apoptosis associated transcript in bladder cancer      AATBC   \n",
       "4                                                NaN        NaN   \n",
       "\n",
       "     type_of_gene      alias  \\\n",
       "0           ncRNA         []   \n",
       "1             NaN         []   \n",
       "2  protein-coding  [C22orf8]   \n",
       "3           ncRNA         []   \n",
       "4             NaN         []   \n",
       "\n",
       "                                             summary notfound ensembl_release  \\\n",
       "0                                                NaN      NaN             109   \n",
       "1                                                NaN      NaN             109   \n",
       "2  Enables identical protein binding activity. Pr...      NaN             109   \n",
       "3                                                NaN      NaN             109   \n",
       "4                                                NaN      NaN             109   \n",
       "\n",
       "  possible_replacement                                          permalink  \n",
       "0                   []  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "1                   []  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "2                   []  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "3                   []  https://feb2023.archive.ensembl.org/Homo_sapie...  \n",
       "4                   []  https://feb2023.archive.ensembl.org/Homo_sapie...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versions = versions[['id', 'release', 'possible_replacement', 'permalink']]\n",
    "versions.rename(columns={'id': 'ensembl_gene_id', 'release': 'ensembl_release'}, inplace=True)\n",
    "\n",
    "gene_table_merged = pd.merge(left = gene_table_merged, right = versions, how = 'left', on = 'ensembl_gene_id')\n",
    "\n",
    "print(gene_table_merged.shape)\n",
    "gene_table_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01eba7c2",
   "metadata": {},
   "source": [
    "### Final cleanup\n",
    "Unfilled \"possible_replacement\" entries should be changed from NaN to empty lists. \n",
    "\n",
    "\"possible_replacement\" entries that have data in them exist as a list of dicts, and need to have the Ensembl IDs pulled out of them as a list of strings. \n",
    "\n",
    "Remove unneeded columns. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d0c07b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in gene_table_merged.loc[gene_table_merged['possible_replacement'].isnull(), 'possible_replacement'].index:\n",
    "    gene_table_merged.at[row, 'possible_replacement'] = []\n",
    "    \n",
    "gene_table_merged['possible_replacement'] = gene_table_merged.apply(\n",
    "    lambda row: row['possible_replacement'] if len(row['possible_replacement']) == 0\n",
    "        else [x['stable_id'] for x in row['possible_replacement']],\n",
    "        axis=1\n",
    ")\n",
    "    \n",
    "gene_table_merged.drop(columns=['index', '_id', '_version', 'notfound'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b371435",
   "metadata": {},
   "source": [
    "### Write to a file\n",
    "This will get uploaded to Synapse as [syn25953363](https://www.synapse.org/#!Synapse:syn25953363)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f2287922",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_table_merged.to_feather('../output/gene_table_merged_GRCh38.p13.feather')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
