{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dd518f3e",
   "metadata": {},
   "source": [
    "# Process SRM differential expression for Agora\n",
    "This script combines SRM differential expression data ([paper](https://doi.org/10.1038%2Fs41591-020-0815-6)) and formats it to be consistent with LFQ and TMT proteomics differential expression data. \n",
    "\n",
    "SRM data was collected in two batches (\"rounds\"). Round 1 tested a single peptide per gene, while round 2 tested multiple peptides for multiple genes. The two rounds were analyzed separately using an ANOVA with post-hoc analysis using TukeyHSD, using the same processing steps as the LFQ and TMT data analysis for Agora. \n",
    "\n",
    "The processed ANOVA data is provided by the researchers on Synapse ([round1](https://www.synapse.org/#!Synapse:syn21448389), [round2](https://www.synapse.org/#!Synapse:syn21448395)), however they did not include confidence intervals or adjusted p-values, which Agora requires. \n",
    "\n",
    "To obtain confidence intervals, we instead ingest the raw data ([round1](https://www.synapse.org/#!Synapse:syn21448381), [round2](https://www.synapse.org/#!Synapse:syn21448378)) and partially replicate their post-hoc analysis (pairwise comparisons using TukeyHSD). The end result produces equivalent log2-fold-changes and p-values to the original processed data but also includes confidence intervals as calculated by TukeyHSD. The code the researchers used for their analysis, which we used for reference, is on Synapse [here](https://www.synapse.org/#!Synapse:syn21448382) and [here](https://www.synapse.org/#!Synapse:syn21448383). \n",
    "\n",
    "This script then adjusts the p-values for each protein using FDR correction, as was done for LFQ and TMT. This script also looks up the Ensembl gene IDs matching each peptide's UniProt ID, and then formats the data frame to match the format of LFQ and TMT data for Agora. For genes with multiple peptides tested, we use the peptide with the smallest corrected p-value in the final data frame, to get a single value per gene. \n",
    "\n",
    "## Install notes\n",
    "This script requires install of several libraries:\n",
    "\n",
    "```\n",
    "pip install openpyxl statsmodels unipressed\n",
    "```\n",
    "\n",
    "and also assumes that Python is running in the environment created by [installing agora-data-tools](https://github.com/Sage-Bionetworks/agora-data-tools#locally)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "328bd339",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.stats.multicomp as mc\n",
    "import statsmodels.stats.multitest as mt\n",
    "from unipressed import IdMappingClient\n",
    "import time\n",
    "import re\n",
    "import synapseclient\n",
    "import agoradatatools.etl.extract as extract\n",
    "import preprocessing_utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "251e9526",
   "metadata": {},
   "source": [
    "# Step 1: Get and clean raw data and metadata\n",
    "The data is downloaded from Synapse, and some gene names are fixed to be consistent across round 1 and round 2. Additionally, two outlier samples are dropped from round 2 data as was done in the [original analysis script](https://www.synapse.org/#!Synapse:syn21448383) for round 2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4ebc848b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.0.0) is available. Your version (3.1.1) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.0.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.Synapse()\n",
    "syn.login(silent=True)\n",
    "\n",
    "round1 = extract.get_entity_as_df(syn_id=\"syn21448381\", source=\"csv\", syn=syn)\n",
    "round1 = round1.rename(columns={\"ProjID\": \"GeneName\"})\n",
    "\n",
    "round2 = extract.get_entity_as_df(syn_id=\"syn21448378\", source=\"csv\", syn=syn)\n",
    "round2 = round2.rename(columns={\"ProjID\": \"GeneName\"})\n",
    "\n",
    "metadata = extract.get_entity_as_df(syn_id=\"syn21448380\", source=\"csv\", syn=syn)\n",
    "metadata = metadata[[\"projid\", \"Group\"]]\n",
    "metadata = metadata.loc[\n",
    "    (metadata[\"Group\"] != \"Exclude\") & (metadata[\"Group\"] != \"SRMcontrol\")\n",
    "]\n",
    "\n",
    "uniprot_mapping = extract.get_entity_as_df(syn_id=\"syn52850588\", source=\"csv\", syn=syn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb86204",
   "metadata": {},
   "source": [
    "Some peptide names are mis-matched between files. This mapping comes from the peptide info spreadsheet and from notes provided by the researchers. Some not listed here are still mis-matched after the re-map but we don't have any more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2032901a",
   "metadata": {},
   "outputs": [],
   "source": [
    "remaps = {\n",
    "    \"bA\": \"APP_3\",\n",
    "    \"bA38\": \"APP_5\",\n",
    "    \"bA42\": \"APP_6\",\n",
    "    \"tau_AT8_s202\": \"MAPT_2\",\n",
    "    \"HLA_B_2\": \"HLA-B_2\",\n",
    "    \"HLA_B_5\": \"HLA-B_5\",\n",
    "}\n",
    "\n",
    "round1[\"GeneName\"] = round1[\"GeneName\"].replace(remaps)\n",
    "round2[\"GeneName\"] = round2[\"GeneName\"].replace(remaps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4712579f",
   "metadata": {},
   "source": [
    "The dataframes are arranged so rows are genes and columns are samples. For the data manipulation below, it is easier for the data to be transposed. Two outlier samples are dropped from round 2 data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c7e61363",
   "metadata": {},
   "outputs": [],
   "source": [
    "round1 = round1.set_index(\"GeneName\").transpose()\n",
    "round2 = round2.set_index(\"GeneName\").transpose()\n",
    "round2 = round2.drop(index=[\"X10300914\", \"X15196262\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af359f25",
   "metadata": {},
   "source": [
    "# Step 2: Get confidence intervals and adjusted p-values\n",
    "\n",
    "Create a function to perform pairwise TukeyHSD for each gene separately. This code is equivalent to and produces identical results to the post-hoc analysis done to produce the [round 1 ANOVA results](https://www.synapse.org/#!Synapse:syn21448389) and [round2 ANOVA results](https://www.synapse.org/#!Synapse:syn21448395), which were originally processed using R scripts for [round 1](https://www.synapse.org/#!Synapse:syn21448382) and [round 2](https://www.synapse.org/#!Synapse:syn21448383). The only difference is that we also include the confidence intervals calculated by `pairwise_tukeyhsd`, which are needed for Agora. \n",
    "\n",
    "We also have to explicitly create a column for the gene name for use in merging later on. In case of testing multiple peptides for a gene, the index/gene will contain an underscore plus a numeric value (e.g `APOD_2`), so we strip off everything after the underscore to get the gene name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2da55947",
   "metadata": {},
   "outputs": [],
   "source": [
    "regex = re.compile(r\"_.*\")\n",
    "\n",
    "\n",
    "def do_tukeyhsd_per_gene(gene, input_df, metadata):\n",
    "    # Merge the metadata and the gene expression column, removing samples with NA data\n",
    "    df = pd.merge(\n",
    "        left=metadata,\n",
    "        right=input_df[gene],\n",
    "        how=\"inner\",\n",
    "        left_on=\"projid\",\n",
    "        right_index=True,\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    "    df = df.dropna()\n",
    "\n",
    "    # The output of pairwise_tukeyhsd doesn't provide an easy way to extract specific comparisons,\n",
    "    # so this puts the output into a pandas DataFrame and pulls out which index corresponds to\n",
    "    # 'AD vs Control'.\n",
    "    result = mc.pairwise_tukeyhsd(endog=df[gene], groups=df[\"Group\"], alpha=0.05)\n",
    "    res_df = pd.DataFrame(result.summary().data, columns=result.summary().data[0])\n",
    "    res_df = res_df.drop(index=0)\n",
    "    ind = np.where((res_df[\"group1\"] == \"AD\") & (res_df[\"group2\"] == \"Control\"))[0][0]\n",
    "\n",
    "    # The diff and confidence interval values need to be opposite sign from what is\n",
    "    # in the result because the result is reporting Control-AD, not AD-Control\n",
    "    results_dict = {\n",
    "        \"GeneName\": regex.sub(\"\", gene),\n",
    "        \"Log2_FC\": -result.meandiffs[ind],\n",
    "        \"CI_Upr\": -result.confint[ind][0],\n",
    "        \"CI_Lwr\": -result.confint[ind][1],\n",
    "        \"PVal\": result.pvalues[ind],\n",
    "    }\n",
    "\n",
    "    return pd.DataFrame(results_dict, index=[gene])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed306325",
   "metadata": {},
   "source": [
    "Run TukeyHSD on each round separately. We do this so we can adjust the p-values for each round separately, since they were treated as two different datasets originally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee8aa80a",
   "metadata": {},
   "outputs": [],
   "source": [
    "round1_list = [do_tukeyhsd_per_gene(gene, round1, metadata) for gene in round1.columns]\n",
    "round1_res = pd.concat(round1_list, axis=0)\n",
    "\n",
    "round2_list = [do_tukeyhsd_per_gene(gene, round2, metadata) for gene in round2.columns]\n",
    "round2_res = pd.concat(round2_list, axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef77459",
   "metadata": {},
   "source": [
    "Adjust p-values for multiple testing *before* combining the data frames, as they were run as separate ANOVAs. Uses FDR correction, which is what was used for LFQ and TMT data processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "89f4e7a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_pvals(df):\n",
    "    (_, adj_p, _, _) = mt.multipletests(pvals=df[\"PVal\"], alpha=0.05, method=\"fdr_bh\")\n",
    "    df[\"Cor_PVal\"] = adj_p\n",
    "    return df\n",
    "\n",
    "\n",
    "round1_res = adjust_pvals(round1_res)\n",
    "round2_res = adjust_pvals(round2_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "692bbd42",
   "metadata": {},
   "source": [
    "Combine round 1 and round 2 into one data frame. Only 4 genes overlap between them, but unlike round 1, which selected a single peptide variant per gene, they ran differential expression on individual peptides and included all peptide variants in the round 2 table. For Agora, we will pick the variant/gene with the smallest corrected p-value between the two rounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "65c12c4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Log2_FC</th>\n",
       "      <th>CI_Upr</th>\n",
       "      <th>CI_Lwr</th>\n",
       "      <th>PVal</th>\n",
       "      <th>Cor_PVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK4</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>0.124460</td>\n",
       "      <td>0.038705</td>\n",
       "      <td>2.639619e-05</td>\n",
       "      <td>1.607768e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>AMPD2</td>\n",
       "      <td>0.145948</td>\n",
       "      <td>0.198572</td>\n",
       "      <td>0.093324</td>\n",
       "      <td>3.555050e-10</td>\n",
       "      <td>6.043584e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>0.077472</td>\n",
       "      <td>-0.120152</td>\n",
       "      <td>8.679973e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP2A2</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.003280</td>\n",
       "      <td>-0.068261</td>\n",
       "      <td>2.674922e-02</td>\n",
       "      <td>6.179993e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>AP2B1</td>\n",
       "      <td>-0.003357</td>\n",
       "      <td>0.039702</td>\n",
       "      <td>-0.046415</td>\n",
       "      <td>9.817129e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>VDAC2</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>0.056417</td>\n",
       "      <td>-0.061966</td>\n",
       "      <td>9.933509e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>VGF</td>\n",
       "      <td>-0.469397</td>\n",
       "      <td>-0.355691</td>\n",
       "      <td>-0.583104</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>VPS35</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.060699</td>\n",
       "      <td>-0.052700</td>\n",
       "      <td>9.850006e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>VSNL1</td>\n",
       "      <td>-0.013745</td>\n",
       "      <td>0.014687</td>\n",
       "      <td>-0.042178</td>\n",
       "      <td>4.928746e-01</td>\n",
       "      <td>7.178826e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>tau</td>\n",
       "      <td>0.613214</td>\n",
       "      <td>0.788548</td>\n",
       "      <td>0.437880</td>\n",
       "      <td>4.668488e-13</td>\n",
       "      <td>1.389536e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneName   Log2_FC    CI_Upr    CI_Lwr          PVal      Cor_PVal\n",
       "0        AK4  0.081583  0.124460  0.038705  2.639619e-05  1.607768e-04\n",
       "67     AMPD2  0.145948  0.198572  0.093324  3.555050e-10  6.043584e-09\n",
       "1    ANKRD40 -0.021340  0.077472 -0.120152  8.679973e-01  9.996966e-01\n",
       "2      AP2A2 -0.035770 -0.003280 -0.068261  2.674922e-02  6.179993e-02\n",
       "68     AP2B1 -0.003357  0.039702 -0.046415  9.817129e-01  9.998368e-01\n",
       "..       ...       ...       ...       ...           ...           ...\n",
       "184    VDAC2 -0.002774  0.056417 -0.061966  9.933509e-01  9.998368e-01\n",
       "64       VGF -0.469397 -0.355691 -0.583104  0.000000e+00  0.000000e+00\n",
       "65     VPS35  0.004000  0.060699 -0.052700  9.850006e-01  9.996966e-01\n",
       "66     VSNL1 -0.013745  0.014687 -0.042178  4.928746e-01  7.178826e-01\n",
       "121      tau  0.613214  0.788548  0.437880  4.668488e-13  1.389536e-11\n",
       "\n",
       "[127 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diffexp = pd.concat([round1_res, round2_res], axis=0, ignore_index=True)\n",
    "rows = diffexp.groupby(\"GeneName\").agg({\"Cor_PVal\": \"idxmin\"})\n",
    "diffexp = diffexp.loc[rows[\"Cor_PVal\"]]\n",
    "\n",
    "diffexp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "141180ef",
   "metadata": {},
   "source": [
    "# Step 3: Get peptide IDs and Ensembl IDs for each gene\n",
    "This step queries UniProt for the Ensembl IDs associated with the UniProt IDs in this dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bea40330",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_info = uniprot_mapping.dropna(subset=[\"UniProtAC\"])\n",
    "uniprot_ids = peptide_info[\"UniProtAC\"].drop_duplicates()\n",
    "request = IdMappingClient.submit(\n",
    "    source=\"UniProtKB_AC-ID\", dest=\"Ensembl\", ids=uniprot_ids\n",
    ")\n",
    "\n",
    "found = False\n",
    "while not found:\n",
    "    time.sleep(2)\n",
    "    try:\n",
    "        ensembl_ids = list(request.each_result())\n",
    "        found = True\n",
    "    except:\n",
    "        print(\"Waiting for response from UniProt...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e42d684",
   "metadata": {},
   "source": [
    "Rename columns to match, and get rid of the Ensembl version at the end of each ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42158477",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_ids = pd.DataFrame(ensembl_ids).rename(\n",
    "    columns={\"from\": \"UniProtAC\", \"to\": \"ENSG\"}\n",
    ")\n",
    "ensembl_ids[\"ENSG\"] = ensembl_ids[\"ENSG\"].str.extract(r\"(\\w+)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eee4be20",
   "metadata": {},
   "source": [
    "Filter out HASGs by querying Biomart for chromosome information and removing genes with HASG chromosomes notation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "df174be3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000005893</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000006125</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000007376</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000008056</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000010310</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>ENSG00000285390</td>\n",
       "      <td>HSCHR1_8_CTG3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>ENSG00000288283</td>\n",
       "      <td>HG109_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>ENSG00000288299</td>\n",
       "      <td>HG2111_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>ENSG00000288301</td>\n",
       "      <td>HG2087_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>ENSG00000288325</td>\n",
       "      <td>HG2087_PATCH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ENSG chromosome_name\n",
       "0    ENSG00000005893               X\n",
       "1    ENSG00000006125              17\n",
       "2    ENSG00000007376              16\n",
       "3    ENSG00000008056               X\n",
       "4    ENSG00000010310              19\n",
       "..               ...             ...\n",
       "246  ENSG00000285390   HSCHR1_8_CTG3\n",
       "247  ENSG00000288283     HG109_PATCH\n",
       "248  ENSG00000288299    HG2111_PATCH\n",
       "249  ENSG00000288301    HG2087_PATCH\n",
       "250  ENSG00000288325    HG2087_PATCH\n",
       "\n",
       "[251 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attributes = [\"ensembl_gene_id\", \"chromosome_name\"]\n",
    "filters = {\"ensembl_gene_id\": set(ensembl_ids[\"ENSG\"])}\n",
    "\n",
    "biomart_result = preprocessing_utils.manual_query_biomart(\n",
    "    attributes=attributes, filters=filters\n",
    ")\n",
    "biomart_result = biomart_result.rename(\n",
    "    columns={\"Gene stable ID\": \"ENSG\", \"Chromosome/scaffold name\": \"chromosome_name\"}\n",
    ")\n",
    "biomart_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6fd2de86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000005893</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000006125</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000007376</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000008056</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000010310</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>ENSG00000234745</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>ENSG00000241468</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>ENSG00000265681</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>ENSG00000267855</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ENSG chromosome_name\n",
       "0    ENSG00000005893               X\n",
       "1    ENSG00000006125              17\n",
       "2    ENSG00000007376              16\n",
       "3    ENSG00000008056               X\n",
       "4    ENSG00000010310              19\n",
       "..               ...             ...\n",
       "224  ENSG00000224051               1\n",
       "225  ENSG00000234745               6\n",
       "226  ENSG00000241468               7\n",
       "227  ENSG00000265681              18\n",
       "228  ENSG00000267855              19\n",
       "\n",
       "[229 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_ids = preprocessing_utils.filter_hasgs(\n",
    "    df=biomart_result, chromosome_name_column=\"chromosome_name\"\n",
    ")\n",
    "valid_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7cffdf9",
   "metadata": {},
   "source": [
    "Update the list of Ensembl IDs to only contain valid (non-HASG) genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e2723421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtAC</th>\n",
       "      <th>ENSG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9UKV3</td>\n",
       "      <td>ENSG00000100813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O94973</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P02649</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P05067</td>\n",
       "      <td>ENSG00000142192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>P45880</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>O95619</td>\n",
       "      <td>ENSG00000127337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>P49750</td>\n",
       "      <td>ENSG00000119596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Q9H0M4</td>\n",
       "      <td>ENSG00000078487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Q95365</td>\n",
       "      <td>ENSG00000234745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProtAC             ENSG\n",
       "0      Q9UKV3  ENSG00000100813\n",
       "1      Q6AI12  ENSG00000154945\n",
       "2      O94973  ENSG00000183020\n",
       "3      P02649  ENSG00000130203\n",
       "4      P05067  ENSG00000142192\n",
       "..        ...              ...\n",
       "249    P45880  ENSG00000165637\n",
       "250    O95619  ENSG00000127337\n",
       "251    P49750  ENSG00000119596\n",
       "252    Q9H0M4  ENSG00000078487\n",
       "253    Q95365  ENSG00000234745\n",
       "\n",
       "[229 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_ids = ensembl_ids.loc[ensembl_ids[\"ENSG\"].isin(valid_ids[\"ENSG\"])]\n",
    "ensembl_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30c01e36",
   "metadata": {},
   "source": [
    "Add the IDs to the peptide info data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ada85f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_info = peptide_info.merge(ensembl_ids, how=\"left\", on=\"UniProtAC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59db8468",
   "metadata": {},
   "source": [
    "Remove unneeded columns from peptide_info and rename remaining columns to match intended output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "762e0863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>ENSG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9UKV3</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>ENSG00000100813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O94973</td>\n",
       "      <td>AP2A2</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P02649</td>\n",
       "      <td>APOE</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P05067</td>\n",
       "      <td>APP</td>\n",
       "      <td>ENSG00000142192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>P45880</td>\n",
       "      <td>VDAC2</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>O95619</td>\n",
       "      <td>YEATS4</td>\n",
       "      <td>ENSG00000127337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>P49750</td>\n",
       "      <td>YLPM1</td>\n",
       "      <td>ENSG00000119596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>Q9H0M4</td>\n",
       "      <td>ZCWPW1</td>\n",
       "      <td>ENSG00000078487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>Q95365</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>ENSG00000234745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProtID GeneName             ENSG\n",
       "0      Q9UKV3    ACIN1  ENSG00000100813\n",
       "2      Q6AI12  ANKRD40  ENSG00000154945\n",
       "4      O94973    AP2A2  ENSG00000183020\n",
       "6      P02649     APOE  ENSG00000130203\n",
       "8      P05067      APP  ENSG00000142192\n",
       "..        ...      ...              ...\n",
       "452    P45880    VDAC2  ENSG00000165637\n",
       "454    O95619   YEATS4  ENSG00000127337\n",
       "456    P49750    YLPM1  ENSG00000119596\n",
       "458    Q9H0M4   ZCWPW1  ENSG00000078487\n",
       "477    Q95365    HLA-B  ENSG00000234745\n",
       "\n",
       "[229 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_info = peptide_info[[\"UniProtAC\", \"Gene\", \"ENSG\"]].drop_duplicates()\n",
    "peptide_info = peptide_info.rename(\n",
    "    columns={\"UniProtAC\": \"UniProtID\", \"Gene\": \"GeneName\"}\n",
    ")\n",
    "peptide_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58eab5ee",
   "metadata": {},
   "source": [
    "# Step 4: Restructure round 1 & 2 data to match other proteomics data\n",
    "Data must have the same columns and format as [the LFQ differential expression data](https://www.synapse.org/#!Synapse:syn18689335).\n",
    "\n",
    "First, merge in the peptide info to get Ensembl IDs and UniProt IDs. This data also needs a tissue column to match LFQ and TMT.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cc09728e",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final = diffexp.merge(peptide_info, how=\"inner\", on=\"GeneName\")\n",
    "diffexp_final[\"Tissue\"] = \"DLPFC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "20ba32e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Log2_FC</th>\n",
       "      <th>CI_Upr</th>\n",
       "      <th>CI_Lwr</th>\n",
       "      <th>PVal</th>\n",
       "      <th>Cor_PVal</th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK4</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>0.124460</td>\n",
       "      <td>0.038705</td>\n",
       "      <td>2.639619e-05</td>\n",
       "      <td>1.607768e-04</td>\n",
       "      <td>P27144</td>\n",
       "      <td>ENSG00000162433</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMPD2</td>\n",
       "      <td>0.145948</td>\n",
       "      <td>0.198572</td>\n",
       "      <td>0.093324</td>\n",
       "      <td>3.555050e-10</td>\n",
       "      <td>6.043584e-09</td>\n",
       "      <td>Q01433</td>\n",
       "      <td>ENSG00000116337</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>0.077472</td>\n",
       "      <td>-0.120152</td>\n",
       "      <td>8.679973e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AP2A2</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.003280</td>\n",
       "      <td>-0.068261</td>\n",
       "      <td>2.674922e-02</td>\n",
       "      <td>6.179993e-02</td>\n",
       "      <td>O94973</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP2B1</td>\n",
       "      <td>-0.003357</td>\n",
       "      <td>0.039702</td>\n",
       "      <td>-0.046415</td>\n",
       "      <td>9.817129e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "      <td>P63010</td>\n",
       "      <td>ENSG00000006125</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>VAT1</td>\n",
       "      <td>0.095231</td>\n",
       "      <td>0.185810</td>\n",
       "      <td>0.004653</td>\n",
       "      <td>3.662795e-02</td>\n",
       "      <td>8.180242e-02</td>\n",
       "      <td>Q99536</td>\n",
       "      <td>ENSG00000108828</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>VDAC2</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>0.056417</td>\n",
       "      <td>-0.061966</td>\n",
       "      <td>9.933509e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "      <td>P45880</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>VGF</td>\n",
       "      <td>-0.469397</td>\n",
       "      <td>-0.355691</td>\n",
       "      <td>-0.583104</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>O15240</td>\n",
       "      <td>ENSG00000128564</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>VPS35</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.060699</td>\n",
       "      <td>-0.052700</td>\n",
       "      <td>9.850006e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "      <td>Q96QK1</td>\n",
       "      <td>ENSG00000069329</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>VSNL1</td>\n",
       "      <td>-0.013745</td>\n",
       "      <td>0.014687</td>\n",
       "      <td>-0.042178</td>\n",
       "      <td>4.928746e-01</td>\n",
       "      <td>7.178826e-01</td>\n",
       "      <td>P62760</td>\n",
       "      <td>ENSG00000163032</td>\n",
       "      <td>DLPFC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneName   Log2_FC    CI_Upr    CI_Lwr          PVal      Cor_PVal  \\\n",
       "0        AK4  0.081583  0.124460  0.038705  2.639619e-05  1.607768e-04   \n",
       "1      AMPD2  0.145948  0.198572  0.093324  3.555050e-10  6.043584e-09   \n",
       "2    ANKRD40 -0.021340  0.077472 -0.120152  8.679973e-01  9.996966e-01   \n",
       "3      AP2A2 -0.035770 -0.003280 -0.068261  2.674922e-02  6.179993e-02   \n",
       "4      AP2B1 -0.003357  0.039702 -0.046415  9.817129e-01  9.998368e-01   \n",
       "..       ...       ...       ...       ...           ...           ...   \n",
       "121     VAT1  0.095231  0.185810  0.004653  3.662795e-02  8.180242e-02   \n",
       "122    VDAC2 -0.002774  0.056417 -0.061966  9.933509e-01  9.998368e-01   \n",
       "123      VGF -0.469397 -0.355691 -0.583104  0.000000e+00  0.000000e+00   \n",
       "124    VPS35  0.004000  0.060699 -0.052700  9.850006e-01  9.996966e-01   \n",
       "125    VSNL1 -0.013745  0.014687 -0.042178  4.928746e-01  7.178826e-01   \n",
       "\n",
       "    UniProtID             ENSG Tissue  \n",
       "0      P27144  ENSG00000162433  DLPFC  \n",
       "1      Q01433  ENSG00000116337  DLPFC  \n",
       "2      Q6AI12  ENSG00000154945  DLPFC  \n",
       "3      O94973  ENSG00000183020  DLPFC  \n",
       "4      P63010  ENSG00000006125  DLPFC  \n",
       "..        ...              ...    ...  \n",
       "121    Q99536  ENSG00000108828  DLPFC  \n",
       "122    P45880  ENSG00000165637  DLPFC  \n",
       "123    O15240  ENSG00000128564  DLPFC  \n",
       "124    Q96QK1  ENSG00000069329  DLPFC  \n",
       "125    P62760  ENSG00000163032  DLPFC  \n",
       "\n",
       "[126 rows x 9 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diffexp_final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c86d9da7",
   "metadata": {},
   "source": [
    "As in the other proteomics data, the unique ID is \"\\<GeneName\\>|\\<UniProtID\\>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "75ad523e",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final[\"UniqID\"] = diffexp_final[\"GeneName\"] + \"|\" + diffexp_final[\"UniProtID\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f47dfdbc",
   "metadata": {},
   "source": [
    "Put the necessary columns in the correct order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "69ca7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final = diffexp_final[\n",
    "    [\n",
    "        \"UniqID\",\n",
    "        \"GeneName\",\n",
    "        \"UniProtID\",\n",
    "        \"ENSG\",\n",
    "        \"Tissue\",\n",
    "        \"Log2_FC\",\n",
    "        \"CI_Upr\",\n",
    "        \"CI_Lwr\",\n",
    "        \"PVal\",\n",
    "        \"Cor_PVal\",\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "50d84968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniqID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Log2_FC</th>\n",
       "      <th>CI_Upr</th>\n",
       "      <th>CI_Lwr</th>\n",
       "      <th>PVal</th>\n",
       "      <th>Cor_PVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK4|P27144</td>\n",
       "      <td>AK4</td>\n",
       "      <td>P27144</td>\n",
       "      <td>ENSG00000162433</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>0.124460</td>\n",
       "      <td>0.038705</td>\n",
       "      <td>2.639619e-05</td>\n",
       "      <td>1.607768e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMPD2|Q01433</td>\n",
       "      <td>AMPD2</td>\n",
       "      <td>Q01433</td>\n",
       "      <td>ENSG00000116337</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.145948</td>\n",
       "      <td>0.198572</td>\n",
       "      <td>0.093324</td>\n",
       "      <td>3.555050e-10</td>\n",
       "      <td>6.043584e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANKRD40|Q6AI12</td>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>0.077472</td>\n",
       "      <td>-0.120152</td>\n",
       "      <td>8.679973e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AP2A2|O94973</td>\n",
       "      <td>AP2A2</td>\n",
       "      <td>O94973</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.003280</td>\n",
       "      <td>-0.068261</td>\n",
       "      <td>2.674922e-02</td>\n",
       "      <td>6.179993e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP2B1|P63010</td>\n",
       "      <td>AP2B1</td>\n",
       "      <td>P63010</td>\n",
       "      <td>ENSG00000006125</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.003357</td>\n",
       "      <td>0.039702</td>\n",
       "      <td>-0.046415</td>\n",
       "      <td>9.817129e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>VAT1|Q99536</td>\n",
       "      <td>VAT1</td>\n",
       "      <td>Q99536</td>\n",
       "      <td>ENSG00000108828</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.095231</td>\n",
       "      <td>0.185810</td>\n",
       "      <td>0.004653</td>\n",
       "      <td>3.662795e-02</td>\n",
       "      <td>8.180242e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>VDAC2|P45880</td>\n",
       "      <td>VDAC2</td>\n",
       "      <td>P45880</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>0.056417</td>\n",
       "      <td>-0.061966</td>\n",
       "      <td>9.933509e-01</td>\n",
       "      <td>9.998368e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>VGF|O15240</td>\n",
       "      <td>VGF</td>\n",
       "      <td>O15240</td>\n",
       "      <td>ENSG00000128564</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.469397</td>\n",
       "      <td>-0.355691</td>\n",
       "      <td>-0.583104</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>VPS35|Q96QK1</td>\n",
       "      <td>VPS35</td>\n",
       "      <td>Q96QK1</td>\n",
       "      <td>ENSG00000069329</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.060699</td>\n",
       "      <td>-0.052700</td>\n",
       "      <td>9.850006e-01</td>\n",
       "      <td>9.996966e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>VSNL1|P62760</td>\n",
       "      <td>VSNL1</td>\n",
       "      <td>P62760</td>\n",
       "      <td>ENSG00000163032</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.013745</td>\n",
       "      <td>0.014687</td>\n",
       "      <td>-0.042178</td>\n",
       "      <td>4.928746e-01</td>\n",
       "      <td>7.178826e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             UniqID GeneName UniProtID             ENSG Tissue   Log2_FC  \\\n",
       "0        AK4|P27144      AK4    P27144  ENSG00000162433  DLPFC  0.081583   \n",
       "1      AMPD2|Q01433    AMPD2    Q01433  ENSG00000116337  DLPFC  0.145948   \n",
       "2    ANKRD40|Q6AI12  ANKRD40    Q6AI12  ENSG00000154945  DLPFC -0.021340   \n",
       "3      AP2A2|O94973    AP2A2    O94973  ENSG00000183020  DLPFC -0.035770   \n",
       "4      AP2B1|P63010    AP2B1    P63010  ENSG00000006125  DLPFC -0.003357   \n",
       "..              ...      ...       ...              ...    ...       ...   \n",
       "121     VAT1|Q99536     VAT1    Q99536  ENSG00000108828  DLPFC  0.095231   \n",
       "122    VDAC2|P45880    VDAC2    P45880  ENSG00000165637  DLPFC -0.002774   \n",
       "123      VGF|O15240      VGF    O15240  ENSG00000128564  DLPFC -0.469397   \n",
       "124    VPS35|Q96QK1    VPS35    Q96QK1  ENSG00000069329  DLPFC  0.004000   \n",
       "125    VSNL1|P62760    VSNL1    P62760  ENSG00000163032  DLPFC -0.013745   \n",
       "\n",
       "       CI_Upr    CI_Lwr          PVal      Cor_PVal  \n",
       "0    0.124460  0.038705  2.639619e-05  1.607768e-04  \n",
       "1    0.198572  0.093324  3.555050e-10  6.043584e-09  \n",
       "2    0.077472 -0.120152  8.679973e-01  9.996966e-01  \n",
       "3   -0.003280 -0.068261  2.674922e-02  6.179993e-02  \n",
       "4    0.039702 -0.046415  9.817129e-01  9.998368e-01  \n",
       "..        ...       ...           ...           ...  \n",
       "121  0.185810  0.004653  3.662795e-02  8.180242e-02  \n",
       "122  0.056417 -0.061966  9.933509e-01  9.998368e-01  \n",
       "123 -0.355691 -0.583104  0.000000e+00  0.000000e+00  \n",
       "124  0.060699 -0.052700  9.850006e-01  9.996966e-01  \n",
       "125  0.014687 -0.042178  4.928746e-01  7.178826e-01  \n",
       "\n",
       "[126 rows x 10 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diffexp_final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feaac13c",
   "metadata": {},
   "source": [
    "Write to csv and upload to Synapse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6fae337a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "##################################################\n",
      " Uploading file to your external S3 storage \n",
      "##################################################\n",
      "\n",
      "INFO: 2024-01-23 15:47:21 | synapseclient_default | \n",
      "##################################################\n",
      " Uploading file to your external S3 storage \n",
      "##################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "diffexp_final.to_csv(\"../../output/SRM_diff_expr.csv\", index=False)\n",
    "file = synapseclient.File(\"../../output/SRM_diff_expr.csv\", parent=\"syn19275585\")  # syn19275585 is the Agora Proteomics Analysis folder\n",
    "file = syn.store(\n",
    "    file,\n",
    "    used=[\n",
    "        \"syn21448381\",\n",
    "        \"syn21448378\",\n",
    "        \"syn21448380\",\n",
    "        \"syn52850588\",\n",
    "        \"https://doi.org/10.1038%2Fs41591-020-0815-6\",\n",
    "    ],\n",
    "    forceVersion=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "agora-data-tools-ywFp1Gf9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
