{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dd518f3e",
   "metadata": {},
   "source": [
    "# Process SRM differential expression for Agora\n",
    "This script combines SRM differential expression data from round 1 and round 2 and formats it to be consistent with LFQ and TMT proteomics differential expression data. \n",
    "\n",
    "This script adjusts the p-values for each protein using FDR correction, and it also looks up the Ensembl gene IDs for each protein.\n",
    "\n",
    "This script requires install of several libraries:\n",
    "\n",
    "```\n",
    "pip install openpyxl statsmodels unipressed\n",
    "```\n",
    "\n",
    "**TODO:** upload the output file to Synapse with the correct provenances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "328bd339",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import statsmodels.stats.multitest as mt\n",
    "from unipressed import IdMappingClient\n",
    "import time\n",
    "import synapseclient\n",
    "import agoradatatools.etl.extract as extract"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "251e9526",
   "metadata": {},
   "source": [
    "# Step 1: Get and combine differential expression data\n",
    "Rounds 1 & 2 differential expression are available on Synapse: https://www.synapse.org/#!Synapse:syn21444847"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ebc848b",
   "metadata": {},
   "outputs": [],
   "source": [
    "syn = synapseclient.Synapse()\n",
    "syn.login(silent=True)\n",
    "\n",
    "round1 = extract.get_entity_as_df(syn_id=\"syn21448389\",\n",
    "                                  source=\"csv\",\n",
    "                                  syn=syn)\n",
    "round1 = round1.rename(columns = {\"Unnamed: 0\": \"GeneName\"})\n",
    "\n",
    "round2 = extract.get_entity_as_df(syn_id=\"syn21448395\",\n",
    "                                  source=\"csv\",\n",
    "                                  syn=syn)\n",
    "round2 = round2.rename(columns = {\"Unnamed: 0\": \"GeneName\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb86204",
   "metadata": {},
   "source": [
    "Some peptide names are mis-matched between files. This mapping comes from the peptide info spreadsheet. Some not listed here are still mis-matched after the re-map but we don't have any more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2032901a",
   "metadata": {},
   "outputs": [],
   "source": [
    "remaps = {\"bA\": \"APP_3\", \"bA38\": \"APP_5\", \"bA42\": \"APP_6\", \"tau_AT8_s202\": \"MAPT_2\", \n",
    "              \"HLA_B_2\": \"HLA-B_2\", \"HLA_B_5\": \"HLA-B_5\"}\n",
    "        \n",
    "round1[\"GeneName\"] = round1[\"GeneName\"].replace(remaps)\n",
    "round2[\"GeneName\"] = round2[\"GeneName\"].replace(remaps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef77459",
   "metadata": {},
   "source": [
    "Adjust p-values for multiple testing *before* combining the data frames, as they were run as separate ANOVAs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "575b0368",
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_pvals(df):\n",
    "    df['PVal'] = df['Control-AD']\n",
    "    (_, adjP, _, _) = mt.multipletests(pvals=df['PVal'], alpha = 0.05, method='fdr_bh')\n",
    "    df['Cor_PVal'] = adjP\n",
    "    return df\n",
    "    \n",
    "round1 = adjust_pvals(round1)\n",
    "round2 = adjust_pvals(round2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "692bbd42",
   "metadata": {},
   "source": [
    "Combine round 1 and round 2. Only 4 genes overlap between them, but round 2 needs a little extra cleaning. Unlike round 1, which selected a single peptide variant per gene, they ran differential expression on individual peptides and included all peptide variants in the round 2 table. Variants are denoted by \\<Gene\\>_\\<#\\>, e.g. BIN1_2.\n",
    "\n",
    "For Agora, we will pick the variant/gene with the smallest corrected p-value between the two rounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65c12c4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>F-Value</th>\n",
       "      <th>Pr(&gt;F)</th>\n",
       "      <th>AsymAD-AD</th>\n",
       "      <th>Control-AD</th>\n",
       "      <th>Control-AsymAD</th>\n",
       "      <th>diff AsymAD-AD</th>\n",
       "      <th>diff Control-AD</th>\n",
       "      <th>diff Control-AsymAD</th>\n",
       "      <th>PVal</th>\n",
       "      <th>Cor_PVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK4</td>\n",
       "      <td>10.797378</td>\n",
       "      <td>2.291591e-05</td>\n",
       "      <td>0.011495</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.090434</td>\n",
       "      <td>-0.042417</td>\n",
       "      <td>-0.081583</td>\n",
       "      <td>-0.039166</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>0.372186</td>\n",
       "      <td>6.893199e-01</td>\n",
       "      <td>0.679559</td>\n",
       "      <td>0.867997</td>\n",
       "      <td>0.984687</td>\n",
       "      <td>0.028536</td>\n",
       "      <td>0.021340</td>\n",
       "      <td>-0.007196</td>\n",
       "      <td>0.867997</td>\n",
       "      <td>0.999697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP2A2</td>\n",
       "      <td>4.318748</td>\n",
       "      <td>1.356259e-02</td>\n",
       "      <td>0.944784</td>\n",
       "      <td>0.026749</td>\n",
       "      <td>0.015119</td>\n",
       "      <td>-0.003595</td>\n",
       "      <td>0.035770</td>\n",
       "      <td>0.039365</td>\n",
       "      <td>0.026749</td>\n",
       "      <td>0.061800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOE</td>\n",
       "      <td>7.220149</td>\n",
       "      <td>7.699608e-04</td>\n",
       "      <td>0.904732</td>\n",
       "      <td>0.002536</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.012679</td>\n",
       "      <td>-0.122649</td>\n",
       "      <td>-0.135328</td>\n",
       "      <td>0.002536</td>\n",
       "      <td>0.008736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APP</td>\n",
       "      <td>635.642124</td>\n",
       "      <td>1.761202e-179</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.079320</td>\n",
       "      <td>-5.757035</td>\n",
       "      <td>-4.677715</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>TPRG1L</td>\n",
       "      <td>7.630556</td>\n",
       "      <td>5.139289e-04</td>\n",
       "      <td>0.007615</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>0.586144</td>\n",
       "      <td>-0.065940</td>\n",
       "      <td>-0.093259</td>\n",
       "      <td>-0.027320</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>0.007593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>UGT8</td>\n",
       "      <td>1.881349</td>\n",
       "      <td>1.529181e-01</td>\n",
       "      <td>0.167573</td>\n",
       "      <td>0.377589</td>\n",
       "      <td>0.991187</td>\n",
       "      <td>-0.123945</td>\n",
       "      <td>-0.112960</td>\n",
       "      <td>0.010985</td>\n",
       "      <td>0.377589</td>\n",
       "      <td>0.580954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>UQCR10</td>\n",
       "      <td>4.399957</td>\n",
       "      <td>1.251457e-02</td>\n",
       "      <td>0.092170</td>\n",
       "      <td>0.017405</td>\n",
       "      <td>0.565345</td>\n",
       "      <td>0.029624</td>\n",
       "      <td>0.047856</td>\n",
       "      <td>0.018233</td>\n",
       "      <td>0.017405</td>\n",
       "      <td>0.049458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>UQCRC2</td>\n",
       "      <td>7.667740</td>\n",
       "      <td>4.954171e-04</td>\n",
       "      <td>0.014425</td>\n",
       "      <td>0.001008</td>\n",
       "      <td>0.396875</td>\n",
       "      <td>0.041283</td>\n",
       "      <td>0.065476</td>\n",
       "      <td>0.024193</td>\n",
       "      <td>0.001008</td>\n",
       "      <td>0.004763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>VDAC2</td>\n",
       "      <td>0.171997</td>\n",
       "      <td>8.420067e-01</td>\n",
       "      <td>0.833434</td>\n",
       "      <td>0.993351</td>\n",
       "      <td>0.935554</td>\n",
       "      <td>0.011739</td>\n",
       "      <td>0.002774</td>\n",
       "      <td>-0.008965</td>\n",
       "      <td>0.993351</td>\n",
       "      <td>0.999837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>164 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneName     F-Value         Pr(>F)  AsymAD-AD  Control-AD  \\\n",
       "0        AK4   10.797378   2.291591e-05   0.011495    0.000026   \n",
       "1    ANKRD40    0.372186   6.893199e-01   0.679559    0.867997   \n",
       "2      AP2A2    4.318748   1.356259e-02   0.944784    0.026749   \n",
       "3       APOE    7.220149   7.699608e-04   0.904732    0.002536   \n",
       "5        APP  635.642124  1.761202e-179   0.000000    0.000000   \n",
       "..       ...         ...            ...        ...         ...   \n",
       "215   TPRG1L    7.630556   5.139289e-04   0.007615    0.001742   \n",
       "216     UGT8    1.881349   1.529181e-01   0.167573    0.377589   \n",
       "218   UQCR10    4.399957   1.251457e-02   0.092170    0.017405   \n",
       "220   UQCRC2    7.667740   4.954171e-04   0.014425    0.001008   \n",
       "221    VDAC2    0.171997   8.420067e-01   0.833434    0.993351   \n",
       "\n",
       "     Control-AsymAD  diff AsymAD-AD  diff Control-AD  diff Control-AsymAD  \\\n",
       "0          0.090434       -0.042417        -0.081583            -0.039166   \n",
       "1          0.984687        0.028536         0.021340            -0.007196   \n",
       "2          0.015119       -0.003595         0.035770             0.039365   \n",
       "3          0.000962        0.012679        -0.122649            -0.135328   \n",
       "5          0.000000       -1.079320        -5.757035            -4.677715   \n",
       "..              ...             ...              ...                  ...   \n",
       "215        0.586144       -0.065940        -0.093259            -0.027320   \n",
       "216        0.991187       -0.123945        -0.112960             0.010985   \n",
       "218        0.565345        0.029624         0.047856             0.018233   \n",
       "220        0.396875        0.041283         0.065476             0.024193   \n",
       "221        0.935554        0.011739         0.002774            -0.008965   \n",
       "\n",
       "         PVal  Cor_PVal  \n",
       "0    0.000026  0.000161  \n",
       "1    0.867997  0.999697  \n",
       "2    0.026749  0.061800  \n",
       "3    0.002536  0.008736  \n",
       "5    0.000000  0.000000  \n",
       "..        ...       ...  \n",
       "215  0.001742  0.007593  \n",
       "216  0.377589  0.580954  \n",
       "218  0.017405  0.049458  \n",
       "220  0.001008  0.004763  \n",
       "221  0.993351  0.999837  \n",
       "\n",
       "[164 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diffexp = pd.concat([round1, round2], axis = 0, ignore_index = True)\n",
    "diffexp[\"GeneName\"] = diffexp[\"GeneName\"].str.replace(r\"_.*\", \"\", regex = True)\n",
    "\n",
    "rows = diffexp.groupby(\"GeneName\").agg({\"Cor_PVal\": \"idxmin\"})\n",
    "diffexp = diffexp.loc[rows[\"Cor_PVal\"].sort_values()]\n",
    "\n",
    "diffexp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "141180ef",
   "metadata": {},
   "source": [
    "# Step 2: Get peptide IDs and Ensembl IDs for each gene\n",
    "Peptide info was provided by the lab via email. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "28005428",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_info = pd.read_excel('../../input/srm/Supplementary Tables R2.xlsx', sheet_name=0)\n",
    "peptide_info = peptide_info.dropna(subset = ['UniProtAC'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68423de6",
   "metadata": {},
   "source": [
    "Get mapping between Uniprot AC and Ensembl ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bea40330",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_ids = peptide_info['UniProtAC'].drop_duplicates()\n",
    "request = IdMappingClient.submit(source='UniProtKB_AC-ID', dest='Ensembl', ids=uniprot_ids)\n",
    "\n",
    "found = False\n",
    "while not found:\n",
    "    time.sleep(2)\n",
    "    try:\n",
    "        ensembl_ids = list(request.each_result())\n",
    "        found = True\n",
    "    except: \n",
    "        print(\"Waiting for response from UniProt...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e42d684",
   "metadata": {},
   "source": [
    "Rename columns to match, and get rid of the Ensembl version at the end of each ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "42158477",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_ids = pd.DataFrame(ensembl_ids).rename(columns = {'from': 'UniProtAC', 'to': 'ENSG'})\n",
    "ensembl_ids['ENSG'] = ensembl_ids['ENSG'].str.extract(r'(\\w+)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30c01e36",
   "metadata": {},
   "source": [
    "Add the IDs to the peptide info data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada85f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_info = peptide_info.merge(ensembl_ids, how = 'left', on = 'UniProtAC')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59db8468",
   "metadata": {},
   "source": [
    "Remove unneeded columns from peptide_info and rename remaining columns to match intended output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "762e0863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>ENSG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9UKV3</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>ENSG00000100813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O94973</td>\n",
       "      <td>AP2A2</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P02649</td>\n",
       "      <td>APOE</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P05067</td>\n",
       "      <td>APP</td>\n",
       "      <td>ENSG00000142192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>P45880</td>\n",
       "      <td>VDAC2</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>O95619</td>\n",
       "      <td>YEATS4</td>\n",
       "      <td>ENSG00000127337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>P49750</td>\n",
       "      <td>YLPM1</td>\n",
       "      <td>ENSG00000119596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>Q9H0M4</td>\n",
       "      <td>ZCWPW1</td>\n",
       "      <td>ENSG00000078487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>Q95365</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>ENSG00000234745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProtID GeneName             ENSG\n",
       "0      Q9UKV3    ACIN1  ENSG00000100813\n",
       "2      Q6AI12  ANKRD40  ENSG00000154945\n",
       "4      O94973    AP2A2  ENSG00000183020\n",
       "6      P02649     APOE  ENSG00000130203\n",
       "8      P05067      APP  ENSG00000142192\n",
       "..        ...      ...              ...\n",
       "510    P45880    VDAC2  ENSG00000165637\n",
       "512    O95619   YEATS4  ENSG00000127337\n",
       "514    P49750    YLPM1  ENSG00000119596\n",
       "516    Q9H0M4   ZCWPW1  ENSG00000078487\n",
       "535    Q95365    HLA-B  ENSG00000234745\n",
       "\n",
       "[254 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_info = peptide_info[['UniProtAC', 'Gene', 'ENSG']].drop_duplicates()\n",
    "peptide_info = peptide_info.rename(columns = {\"UniProtAC\": \"UniProtID\", \"Gene\": \"GeneName\"})\n",
    "peptide_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58eab5ee",
   "metadata": {},
   "source": [
    "# Step 4: Restructure round 1 & 2 data to match other proteomics data\n",
    "Data must have the same columns and format as https://www.synapse.org/#!Synapse:syn18689335\n",
    "\n",
    "First, create the same columns that the other proteomics data has:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "08e35fdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make log fold-change values negative/opposite what's in the original data because the comparison in the original data\n",
    "# is control - AD rather than AD - control, which is what the other proteomics datasets use. \n",
    "diffexp['Log2_FC'] = -diffexp['diff Control-AD']\n",
    "diffexp['PVal'] = -diffexp['Control-AD']\n",
    "diffexp['Tissue'] = 'DLPFC'\n",
    "\n",
    "# Fake values since we don't have this info from the file. \n",
    "# CI_Upr and CI_Lwr are not used in Agora but the transform doesn't allow NA values. \n",
    "# We could get these values by re-calculating the ANOVAs if we need to. \n",
    "diffexp['CI_Upr'] = diffexp['Log2_FC'] \n",
    "diffexp['CI_Lwr'] = diffexp['Log2_FC']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ee5e0b",
   "metadata": {},
   "source": [
    "Merge in the peptide info to get Ensembl IDs and UniProt IDs. Use an inner join to get rid of the \"averaged\" peptide rows that are in the round 2 data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc09728e",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final = diffexp.merge(peptide_info, how = 'inner', on = 'GeneName')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c86d9da7",
   "metadata": {},
   "source": [
    "As in the other proteomics data, the unique ID is \"\\<GeneName\\>|\\<UniProtID\\>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "75ad523e",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final[\"UniqID\"] = diffexp_final[\"GeneName\"] + \"|\" + diffexp_final[\"UniProtID\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f47dfdbc",
   "metadata": {},
   "source": [
    "Put the necessary columns in the correct order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "69ca7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final = diffexp_final[[\"UniqID\", \"GeneName\", \"UniProtID\", \"ENSG\", \"Tissue\", \n",
    "                               \"Log2_FC\", \"CI_Upr\", \"CI_Lwr\", \"PVal\", \"Cor_PVal\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "50d84968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniqID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Log2_FC</th>\n",
       "      <th>CI_Upr</th>\n",
       "      <th>CI_Lwr</th>\n",
       "      <th>PVal</th>\n",
       "      <th>Cor_PVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AK4|P27144</td>\n",
       "      <td>AK4</td>\n",
       "      <td>P27144</td>\n",
       "      <td>ENSG00000162433</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>0.081583</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKRD40|Q6AI12</td>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>ENSG00000154945</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>-0.021340</td>\n",
       "      <td>-0.867997</td>\n",
       "      <td>0.999697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP2A2|O94973</td>\n",
       "      <td>AP2A2</td>\n",
       "      <td>O94973</td>\n",
       "      <td>ENSG00000183020</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.035770</td>\n",
       "      <td>-0.026749</td>\n",
       "      <td>0.061800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOE|P02649</td>\n",
       "      <td>APOE</td>\n",
       "      <td>P02649</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.122649</td>\n",
       "      <td>0.122649</td>\n",
       "      <td>0.122649</td>\n",
       "      <td>-0.002536</td>\n",
       "      <td>0.008736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APP|P05067</td>\n",
       "      <td>APP</td>\n",
       "      <td>P05067</td>\n",
       "      <td>ENSG00000142192</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>5.757035</td>\n",
       "      <td>5.757035</td>\n",
       "      <td>5.757035</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>TPRG1L|Q5T0D9</td>\n",
       "      <td>TPRG1L</td>\n",
       "      <td>Q5T0D9</td>\n",
       "      <td>ENSG00000158109</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.093259</td>\n",
       "      <td>0.093259</td>\n",
       "      <td>0.093259</td>\n",
       "      <td>-0.001742</td>\n",
       "      <td>0.007593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>UGT8|Q16880</td>\n",
       "      <td>UGT8</td>\n",
       "      <td>Q16880</td>\n",
       "      <td>ENSG00000174607</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>0.112960</td>\n",
       "      <td>0.112960</td>\n",
       "      <td>0.112960</td>\n",
       "      <td>-0.377589</td>\n",
       "      <td>0.580954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>UQCR10|Q9UDW1</td>\n",
       "      <td>UQCR10</td>\n",
       "      <td>Q9UDW1</td>\n",
       "      <td>ENSG00000184076</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.047856</td>\n",
       "      <td>-0.047856</td>\n",
       "      <td>-0.047856</td>\n",
       "      <td>-0.017405</td>\n",
       "      <td>0.049458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>UQCRC2|P22695</td>\n",
       "      <td>UQCRC2</td>\n",
       "      <td>P22695</td>\n",
       "      <td>ENSG00000140740</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.065476</td>\n",
       "      <td>-0.065476</td>\n",
       "      <td>-0.065476</td>\n",
       "      <td>-0.001008</td>\n",
       "      <td>0.004763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>VDAC2|P45880</td>\n",
       "      <td>VDAC2</td>\n",
       "      <td>P45880</td>\n",
       "      <td>ENSG00000165637</td>\n",
       "      <td>DLPFC</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>-0.002774</td>\n",
       "      <td>-0.993351</td>\n",
       "      <td>0.999837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             UniqID GeneName UniProtID             ENSG Tissue   Log2_FC  \\\n",
       "0        AK4|P27144      AK4    P27144  ENSG00000162433  DLPFC  0.081583   \n",
       "1    ANKRD40|Q6AI12  ANKRD40    Q6AI12  ENSG00000154945  DLPFC -0.021340   \n",
       "2      AP2A2|O94973    AP2A2    O94973  ENSG00000183020  DLPFC -0.035770   \n",
       "3       APOE|P02649     APOE    P02649  ENSG00000130203  DLPFC  0.122649   \n",
       "4        APP|P05067      APP    P05067  ENSG00000142192  DLPFC  5.757035   \n",
       "..              ...      ...       ...              ...    ...       ...   \n",
       "132   TPRG1L|Q5T0D9   TPRG1L    Q5T0D9  ENSG00000158109  DLPFC  0.093259   \n",
       "133     UGT8|Q16880     UGT8    Q16880  ENSG00000174607  DLPFC  0.112960   \n",
       "134   UQCR10|Q9UDW1   UQCR10    Q9UDW1  ENSG00000184076  DLPFC -0.047856   \n",
       "135   UQCRC2|P22695   UQCRC2    P22695  ENSG00000140740  DLPFC -0.065476   \n",
       "136    VDAC2|P45880    VDAC2    P45880  ENSG00000165637  DLPFC -0.002774   \n",
       "\n",
       "       CI_Upr    CI_Lwr      PVal  Cor_PVal  \n",
       "0    0.081583  0.081583 -0.000026  0.000161  \n",
       "1   -0.021340 -0.021340 -0.867997  0.999697  \n",
       "2   -0.035770 -0.035770 -0.026749  0.061800  \n",
       "3    0.122649  0.122649 -0.002536  0.008736  \n",
       "4    5.757035  5.757035 -0.000000  0.000000  \n",
       "..        ...       ...       ...       ...  \n",
       "132  0.093259  0.093259 -0.001742  0.007593  \n",
       "133  0.112960  0.112960 -0.377589  0.580954  \n",
       "134 -0.047856 -0.047856 -0.017405  0.049458  \n",
       "135 -0.065476 -0.065476 -0.001008  0.004763  \n",
       "136 -0.002774 -0.002774 -0.993351  0.999837  \n",
       "\n",
       "[137 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diffexp_final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feaac13c",
   "metadata": {},
   "source": [
    "Write to csv and upload to Synapse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fae337a",
   "metadata": {},
   "outputs": [],
   "source": [
    "diffexp_final.to_csv('../../output/SRM_diff_expr.csv', index = False)\n",
    "file = synapseclient.File('../../output/SRM_diff_expr.csv', parent = 'syn7525089') # syn7525089 is the Agora Raw Data folder\n",
    "file = syn.store(file, used = ['syn21448389', 'syn21448395'], forceVersion=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "agora-data-tools-ywFp1Gf9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
